<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html> <head>
     <title>Apophenia: a library for scientific computing</title>
<!-- Google is watching. -->
<script src="http://www.google-analytics.com/urchin.js" type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-134313-2";
urchinTracker();
</script>
     <link rel="stylesheet" href="typical.css" type="text/css" >
     <script  type="text/javascript" language="JavaScript" src="tree.js"></script>
</head><body>
     <center><table cellpadding=10pt>
     <tr> <td><img width=140px src=flake.gif alt="Patterns in static"></td> 
    <td><table>
     	<tr> <td><center><h2><a href="http://apophenia.info">Apophenia</a></h2></center></td></tr>
<tr><td><div class="qindex"><a class="qindex" href="index.html">&nbsp;Intro</a> | <a class="qindex" href="outline.html">Outline</a> | <a class="qindex" href="globals.html">Index</a> <!--| <a class="qindex" href="files.html">File&nbsp;List&nbsp;</a> -->  </div></td></tr></table>
	</td></tr></table></center>
    <div> <!--Doxygen generates an extra </div>.-->
<!-- Generated by Doxygen 1.8.9.1 -->
</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">An outline of the library </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><span class="trigger" onClick="showAll();"<a>Expand all </a></span> | <span class="trigger" onClick="closeAll();"<a>Collapse all </a></span></p>
<p><h2><a class="anchor" name="preliminaries"><div class="trigger" onClick="showBranch('preliminariesd');swapFolder('preliminariesf')"><img src="right.png" border="0" id="preliminariesf" alt="pip"> Getting started</div></a></h2><div class="branch" id="preliminariesd">
<p>If you are entirely new to Apophenia, <a class="el" href="gentle.html">have a look at the Gentle Introduction here</a>.</p>
<p>As well as the information in this outline, there is a separate page covering the details of <a class="el" href="setup.html">setting up a computing environment</a> and another page with <a class="el" href="eg.html">some sample code</a> for your perusal.</p>
<p>For another concrete example of folding Apophenia into a project, have a look at this <a class="el" href="setup.html#sample_program">sample program</a>.</p>
<p><h2><a class="anchor" name="c"><div class="trigger" onClick="showBranch('cd');swapFolder('cf')"><img src="right.png" border="0" id="cf" alt="pip"> Some notes on C and Apophenia's use of C utilities.</div></a></h2><div class="branch" id="cd">
<p><h2><a class="anchor" name="learning"><div class="trigger" onClick="showBranch('learningd');swapFolder('learningf')"><img src="right.png" border="0" id="learningf" alt="pip"> Learning C</div></a></h2><div class="branch" id="learningd">
<p><a href="http://modelingwithdata.org">Modeling with Data</a> has a full tutorial for C, oriented at users of standard stats packages. More nuts-and-bolts tutorials are <a href="http://www.google.com/search?hl=es&amp;amp;c2coff=1&amp;amp;q=c+tutorial">in abundance</a>. Some people find pointers to be especially difficult; fortunately, there's a <a href="http://www.youtube.com/watch?v=6pmWojisM_E">claymation cartoon</a> which clarifies everything.</p>
<p>Coding often relies on gathering together many libraries; there is a section at the bottom of this outline linking to references for some libraries upon which Apophenia builds.</p>
<p></div>
<p><h2><a class="anchor" name="usagenotes"><div class="trigger" onClick="showBranch('usagenotesd');swapFolder('usagenotesf')"><img src="right.png" border="0" id="usagenotesf" alt="pip"> Usage notes</div></a></h2><div class="branch" id="usagenotesd">
<p>Here are some notes about the technical details of using the Apophenia library in your development environment.</p>
<p><b> Header aggregation </b></p>
<p>There is only one header. Put </p><div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="apop_8h.html">apop.h</a>&gt;</span></div>
</div><!-- fragment --><p> at the top of your file, and you're done. Everything declared in that file starts with <code>apop_</code> (or <code>Apop_</code>).</p>
<p><b>Linking</b></p>
<p>You will need to link to the Apophenia library, which involves adding the <code>-lapophenia</code> flag to your compiler. Apophenia depends on SQLite3 and the GNU Scientific Library (which depends on a BLAS), so you will probably need something like:</p>
<div class="fragment"><div class="line">gcc sample.c -lapophenia -lsqlite3 -lgsl -lgslcblas -o run_me -g -Wall -O3</div>
</div><!-- fragment --><p>Your best bet is to encapsulate this mess in a <a class="el" href="makefile.html">Makefile</a>. Even if you are using an IDE and its command-line management tools, see the Makefile page for notes on useful flags.</p>
<p><b>Debugging</b> The end of <a href="http://modelingwithdata.org/appendix_o.html">Appendix O</a> of <em>Modeling with Data</em> offers some GDB macros which can make dealing with Apophenia from the GDB command line much more pleasant. As per the next section, it also helps to set <code>apop_opts.stop_on_warning='v'</code> or <code>'w'</code> when running under the debugger.</p>
<p><b>Standards compliance</b></p>
<p>To the best of our abilities, Apophenia complies to the C standard (ISO/IEC 9899:2011).</p>
<p><b> Easier calling syntax</b></p>
<p>Many functions allow optional named arguments to functions. For example: </p><div class="fragment"><div class="line"><a class="code" href="group__convenience__fns.html#ga44a77f5e6867c627fd313f3f6e252ff8">apop_vector_distance</a>(v1, v2); <span class="comment">//assumes Euclidean distance</span></div>
<div class="line"><a class="code" href="group__convenience__fns.html#ga44a77f5e6867c627fd313f3f6e252ff8">apop_vector_distance</a>(v1, .metric=<span class="charliteral">&#39;M&#39;</span>); <span class="comment">//assumes v2=0, uses Manhattan metric.</span></div>
</div><!-- fragment --><p>See the <a class="el" href="designated.html">Designated initializers</a> page for details of this syntax.</p>
<p></div>
<p></div>
<p><h2><a class="anchor" name="debugging"><div class="trigger" onClick="showBranch('debuggingd');swapFolder('debuggingf')"><img src="right.png" border="0" id="debuggingf" alt="pip"> Errors, logging, debugging and stopping</div></a></h2><div class="branch" id="debuggingd">
<h5>The <code>error</code> element</h5>
<p>The <a class="el" href="gentle.html#apop_data">apop_data</a> set and the <a class="el" href="structapop__model.html">apop_model</a> both include an element named <code>error</code>. It is normally <code>0</code>, indicating no (known) error.</p>
<p>For example, <a class="el" href="apop__data_8c.html#add67e6838ce2521a1d756894dc9b5ccf">apop_data_copy</a> detects allocation errors and some circular links (when <code>Data-&gt;more == Data</code>) and fails in those cases. You could thus use the function with a form like</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *d = <a class="code" href="group__conversions.html#ga630de5d4fcbe9ef13373029f5d813a97">apop_text_to_data</a>(<span class="stringliteral">&quot;indata&quot;</span>);</div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *cp = <a class="code" href="apop_8h.html#add67e6838ce2521a1d756894dc9b5ccf">apop_data_copy</a>(d);</div>
<div class="line"><span class="keywordflow">if</span> (cp-&gt;error) {printf(<span class="stringliteral">&quot;Couldn&#39;t copy the input data; failing.\n&quot;</span>); <span class="keywordflow">return</span> 1;}</div>
</div><!-- fragment --><p>There is sometimes (but not always) benefit to handling specific error codes, which are listed in the documentation of those functions that set the <code>error</code> element. E.g.,</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *d = <a class="code" href="group__conversions.html#ga630de5d4fcbe9ef13373029f5d813a97">apop_text_to_data</a>(<span class="stringliteral">&quot;indata&quot;</span>);</div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *cp = <a class="code" href="apop_8h.html#add67e6838ce2521a1d756894dc9b5ccf">apop_data_copy</a>(d);</div>
<div class="line"><span class="keywordflow">if</span> (cp-&gt;error == <span class="charliteral">&#39;a&#39;</span>) {printf(<span class="stringliteral">&quot;Couldn&#39;t allocate space for the copy; failing.\n&quot;</span>); <span class="keywordflow">return</span> 1;}</div>
<div class="line"><span class="keywordflow">if</span> (cp-&gt;error == <span class="charliteral">&#39;c&#39;</span>) {printf(<span class="stringliteral">&quot;Circular link in the data set; failing.\n&quot;</span>); <span class="keywordflow">return</span> 2;}</div>
</div><!-- fragment --><h5>Verbosity level and logging</h5>
<p>The global variable <code>apop_opts.verbose</code> determines how many notifications and warnings get printed by Apophenia's warning mechanism:</p>
<p>-1: turn off logging, print nothing (ill-advised) <br />
 0: notify only of failures and clear danger <br />
 1: warn of technically correct but odd situations that might indicate, e.g., numeric instability <br />
 2: debugging-type information; print queries <br />
 3: give me everything, such as the state of the data at each iteration of a loop.</p>
<p>These levels are of course subjective, but should give you some idea of where to place the verbosity level. The default is 1.</p>
<p>The messages are printed to the <code>FILE</code> handle at <code>apop_opts.log_file</code>. If this is blank (which happens at startup), then this is set to <code>stderr</code>. This is the typical behavior for a console program. Use</p>
<div class="fragment"><div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.<a class="code" href="structapop__opts__type.html#a27e596063680b5b56ba5bf39fc2b6647">log_file</a> = fopen(<span class="stringliteral">&quot;mylog&quot;</span>, <span class="stringliteral">&quot;w&quot;</span>);</div>
</div><!-- fragment --><p>to write to the <code>mylog</code> file instead of <code>stderr</code>.</p>
<p>As well as the error and warning messages, some functions can also print diagnostics, using the <a class="el" href="apop_8h.html#a674b73c9b8813f35e41009428cbc80f9">Apop_notify</a> macro. For example, <a class="el" href="group__queries.html#ga236452b3305ea7b791c781d91a819e3b">apop_query</a> and friends will print the query sent to the database engine iff <code>apop_opts.verbose &gt;=2</code> (which is useful when building complex queries). The diagnostics attempt to follow the same verbosity scale as the warning messages.</p>
<h5>Stopping</h5>
<p>Warnings and errors never halt processing. It is up to the calling function to decide whether to stop.</p>
<p>When running the program under a debugger, this is an annoyance: we want to stop as soon as a problem turns up.</p>
<p>The global variable <code>apop_opts.stop_on_warning</code> changes when the system halts:</p>
<p><code>'n'</code>: never halt. If you were using Apophenia to support a user-friendly GUI, for example, you would use this mode.<br />
 The default: if the variable is <code>'\0'</code> (the default), halt on severe errors, continue on all warnings.<br />
 <code>'v'</code>: If the verbosity level of the warning is such that the warning would print to screen, then halt; if the warning message would be filtered out by your verbosity level, continue.<br />
 <code>'w'</code>: Halt on all errors or warnings, including those below your verbosity threshold.</p>
<p>See the documentation for individual functions for details on how each reports errors to the caller and the level at which warnings are posted.</p>
<p></div>
<p><h2><a class="anchor" name="About"><div class="trigger" onClick="showBranch('Aboutd');swapFolder('Aboutf')"><img src="right.png" border="0" id="Aboutf" alt="pip"> SQL, the syntax for querying databases</div></a></h2><div class="branch" id="Aboutd">
<p>For a reference, your best bet is the <a href="http://www.sqlite.org/lang.html">Structured Query Language reference</a> for SQLite. For a tutorial; there is an abundance of <a href="http://www.google.com/search?q=sql+tutorial">tutorials online</a>. Here is a nice blog <a href="http://fluff.info/blog/arch/00000118.htm">entry</a> about complementaries between SQL and matrix manipulation packages.</p>
<p>Apophenia currently supports two database engines: SQLite and mySQL/mariaDB. SQLite is the default, because it is simpler and generally more easygoing than mySQL, and supports in-memory databases.</p>
<p>The global <code>apop_opts.db_engine</code> is initially <code>NUL</code>, indicating no preference for a database engine. You can explicitly set it:</p>
<div class="fragment"><div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.<a class="code" href="structapop__opts__type.html#a9c128a43c2e53c7639ce4ddb0cef92f6">db_engine</a>=<span class="charliteral">&#39;s&#39;</span> <span class="comment">//use SQLite</span></div>
<div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.<a class="code" href="structapop__opts__type.html#a9c128a43c2e53c7639ce4ddb0cef92f6">db_engine</a>=<span class="charliteral">&#39;m&#39;</span> <span class="comment">//use mySQL/mariaDB</span></div>
</div><!-- fragment --><p>If <code>apop_opts.db_engine</code> is still <code>NUL</code> on your first database operation, then I will check for an environment variable <code>APOP_DB_ENGINE</code>, and set <code>apop_opts.db_engine='m'</code> if it is found and matches (case insensitive) <code>mariadb</code> or <code>mysql</code>.</p>
<div class="fragment"><div class="line">export APOP_DB_ENGINE=mariadb</div>
<div class="line"><a class="code" href="group__conversions.html#gabd00e16791a970631c3ed4ac3487da46">apop_text_to_db</a> indata mtab db_for_maria</div>
<div class="line"></div>
<div class="line">unset APOP_DB_ENGINE</div>
<div class="line"><a class="code" href="group__conversions.html#gabd00e16791a970631c3ed4ac3487da46">apop_text_to_db</a> indata stab db_for_sqlite.db</div>
</div><!-- fragment --><p>Finally, Apophenia provides a few nonstandard SQL functions to facilitate math via database; see <a class="el" href="db_moments.html">Database moments (plus pow()!)</a>. </div>
<p><h2><a class="anchor" name="threads"><div class="trigger" onClick="showBranch('threadsd');swapFolder('threadsf')"><img src="right.png" border="0" id="threadsf" alt="pip"> Threading</div></a></h2><div class="branch" id="threadsd">
<p>Apophenia uses OpenMP for threading. You generally do not need to know how OpenMP works to use Apophenia, and many points of work will thread without your doing anything.</p>
<ul>
<li>All functions strive to be thread-safe. Part of how this is achieved is that static variables are marked as thread-local or atomic, as per the C standard. There still exist compilers that can't implement thread-local or atomic variables, in which case your safest bet is to set OMP's thread count to one as below (or get a new compiler).</li>
</ul>
<ul>
<li>Some functions modify their inputs. It is up to you to use those functions in a thread-safe manner. The <a class="el" href="apop__data_8c.html#aa06b6ab6ec9f78f305b1dacc8ec2efd3">apop_matrix_realloc</a> handles states and global variables correctly in a threaded environment, but if you have two threads resizing the same <code>gsl_matrix</code> at the same time, you're going to have problems.</li>
</ul>
<ul>
<li>There are few compilers that don't support OpenMP. Clang on MacOS may be the only current mainstream example as of this writing, and they are hard at work on implementing it. In the mean time, when compiling on such a system all work will be single-threaded.</li>
</ul>
<ul>
<li>Set the maximum number of threads to <code>N</code> with the environment variable</li>
</ul>
<div class="fragment"><div class="line">export OMP_NUM_THREADS N</div>
</div><!-- fragment --><p>or the C function</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;omp.h&gt;</span></div>
<div class="line">omp_set_num_threads(N);</div>
</div><!-- fragment --><p>Use one of these methods with <code>N=1</code> if you want a single-threaded program.</p>
<ul>
<li><a class="el" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a> and friends distribute their <code>for</code> loop over the input <a class="el" href="gentle.html#apop_data">apop_data</a> set across multiple threads. Therefore, be careful to send thread-unsafe functions to it only after calling <code>omp_set_num_threads(1)</code>.</li>
</ul>
<ul>
<li>There are a few functions, like <a class="el" href="apop__asst_8c.html#a4e204cdc7429bef3ca2efb83453fe77a">apop_model_draws</a>, that rely on <a class="el" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a>, and therefore also thread by default.</li>
</ul>
<ul>
<li>The function <a class="el" href="apop_8h.html#a1e829ce2396bf05ab10460bcc37aeeab">apop_rng_get_thread</a> retrieves a statically-stored RNG specific to a given thread. Therefore, if you use that function in the place of a <code>gsl_rng</code>, you can parallelize functions that make random draws.</li>
</ul>
<ul>
<li><a class="el" href="apop_8h.html#a1e829ce2396bf05ab10460bcc37aeeab">apop_rng_get_thread</a> allocates its store of threads using <code>apop_opts.rng_seed</code>, then incrementing that seed by one. You thus probably have threads with seeds 479901, 479902, 479903, .... [If you have a better way to do it, please feel free to modify the code to implement your improvement and submit a pull request on Github.]</li>
</ul>
<p>See <a href="http://modelingwithdata.org/arch/00000175.htm">this tutorial on C threading</a> if you would like to know more, or are unsure about whether your functions are thread-safe or not.</p>
<p></div>
<p><h2><a class="anchor" name="mwd"><div class="trigger" onClick="showBranch('mwdd');swapFolder('mwdf')"><img src="right.png" border="0" id="mwdf" alt="pip"> The book version</div></a></h2><div class="branch" id="mwdd">
<p>Apophenia co-evolved with <em>Modeling with Data: Tools and Techniques for Statistical Computing</em>. You can read about the book, or download a free PDF copy of the full text, at <a href="http://modelingwithdata.org">modelingwithdata.org</a>.</p>
<p>If you are at this site, there is probably something there for you, including a tutorial on C and general computing form, SQL for data-handing, several chapters of statistics from various perspectives, and more details on working Apophenia.</p>
<p>As with many computer programs, the preferred manner of citing Apophenia is to cite its related book. Here is a BibTeX-formatted entry giving the relevant information:</p>
<div class="fragment"><div class="line">@book{klemens:modeling,</div>
<div class="line">    title = <span class="stringliteral">&quot;Modeling with Data: Tools and Techniques for Statistical Computing&quot;</span>,</div>
<div class="line">    author=<span class="stringliteral">&quot;Ben Klemens&quot;</span>,</div>
<div class="line">    year=2008,</div>
<div class="line">    publisher=<span class="stringliteral">&quot;Princeton University Press&quot;</span></div>
<div class="line">}</div>
</div><!-- fragment --><p>The rationale for the <a class="el" href="structapop__model.html">apop_model</a> struct, based on an algebraic system of models, is detailed in a <a href="http://www.census.gov/srd/papers/pdf/rrs2014-06.pdf">U.S. Census Bureau research report</a>.</p>
<p></div>
<p><h2><a class="anchor" name="status"><div class="trigger" onClick="showBranch('statusd');swapFolder('statusf')"><img src="right.png" border="0" id="statusf" alt="pip"> What is the status of the code?</div></a></h2><div class="branch" id="statusd">
<p>[This section last updated 3 August 2014.]</p>
<p>Apophenia was first posted to SourceForge in February 2005, which means that we've had several years to develop and test the code in real-world applications.</p>
<p>The test suite, including the sample code and solution set for <em>Modeling with Data</em>, is about 5,500 lines over 135 files. gprof reports that it covers over 90% of the 7,700 lines in Apophenia's code base. A broad rule of thumb for any code base is that the well-worn parts, in this case functions like <a class="el" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a> and <a class="el" href="group__models.html#ga297c2dc8dbdf4ae93fc7e10860400466">apop_normal</a>'s <code>log_likelihood</code>, are likely to be entirely reliable, while the out-of-the-way functions (maybe the score for the Beta distribution) are worth a bit of caution. Close to all of the code has been used in production, so all of it was at least initially tested against real-world data.</p>
<p>It is currently at version 0.999, which is intended to indicate that it is substantially complete. Of course, a library for scientific computing, or even for that small subset that is statistics, will never cover all needs and all methods. But as it stands Apophenia's framework, based on the <a class="el" href="gentle.html#apop_data">apop_data</a> and <a class="el" href="structapop__model.html">apop_model</a>, is basically internally consistent, has enough tools that you can get common work done quickly, and is reasonably fleshed out with a good number of models out of the box.</p>
<p>The <a class="el" href="gentle.html#apop_data">apop_data</a> structure is set, and there are enough functions there that you could use it as a subpackage by itself (especially in tandem with the database functions) for nontrivial dealings with data.</p>
<p>The <a class="el" href="structapop__model.html">apop_model</a> structure is much more ambitious&mdash;Apophenia is really intended to be a novel system for developing models&mdash;and its internals can still be improved. The promise underlying the structure is that you can provide just one item, such as an RNG or a likelihood function, and the structure will do all of the work to fill in computationally-intensive methods for everything else; see <a class="el" href="modeldetails.html#settingswriting">Writing new settings groups</a> for the details. Some directions aren't quite there yet (such as RNG -&gt; most other things), the PMF model needs an internal index for faster lookups, and so on. Readers are invited to contribute better methods (such as an alternate means of estimating mixture models), or filling in more of existing models (write a dlog likelihood function for a model that does not currently have one), or submit new standard models not yet included.</p>
<p></div>
<p><h2><a class="anchor" name="ext"><div class="trigger" onClick="showBranch('extd');swapFolder('extf')"><img src="right.png" border="0" id="extf" alt="pip"> How do I write extensions?</div></a></h2><div class="branch" id="extd">
<p>It's not a package, so you don't need an API&mdash;write your code and <code>include</code> it like any other C code. The system is written to not require a registration or initialization step to add a new model or other such parts. A new <a class="el" href="structapop__model.html">apop_model</a> has to conform to some rules if it is to play well with <a class="el" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>, <a class="el" href="group__models.html#ga87f15cd1923ed7b5b46e52a77ca1dfef">apop_draw</a>, and so forth. See the notes at <a class="el" href="modeldetails.html">Writing new models</a>. Once your new model or function is working, please post the code or a link to the code on the <a href="https://github.com/b-k/apophenia/wiki">Apophenia wiki</a>.</p>
<p></div>
<p></div>
<p><h2><a class="anchor" name="dataoverview"><div class="trigger" onClick="showBranch('dataoverviewd');swapFolder('dataoverviewf')"><img src="right.png" border="0" id="dataoverviewf" alt="pip"> Data sets</div></a></h2><div class="branch" id="dataoverviewd">
<p>The <a class="el" href="gentle.html#apop_data">apop_data</a> structure represents a data set. It joins together a <code>gsl_vector</code>, a <code>gsl_matrix</code>, an <a class="el" href="structapop__name.html">apop_name</a>, and a table of strings. It tries to be lightweight, so you can use it everywhere you would use a <code>gsl_matrix</code> or a <code>gsl_vector</code>.</p>
<p>If you are viewing the HTML documentation, here is a diagram showing a sample data set with all of the elements in place. Together, they represet a data set where each row is an observation, which includes both numeric and text values, and where each row/column may be named.</p>
<table frame=box>
<tr>
<td>Rowname</td><td>Vector</td><td> Matrix</td><td> Text</td><td>Weights</td>
</tr><tr valign=bottom>
<td align=center>
<table frame=box>
<tr><td> </td></tr>
<tr>
<td>"Steven"</td>
</tr><tr>
<td>"Sandra"</td>
</tr><tr>
<td>"Joe"</td><td>
</tr> 
</table>
</td><td align=center>
<table frame=box>
<tr>
<th>Outcome</th>
</tr> <tr>
<td align=center>1</td>
</tr><tr>
<td align=center>0</td>
</tr><tr>
<td align=center>1</td>
</tr> 
</table>
</td><td align=center>
<table frame=box>
<tr>
<th> Age</th><th> Weight (kg)</th><th> Height (cm)</th>
</tr> <tr>
<td> 32</td><td> 65</td><td> 175</td>
</tr><tr>
<td> 41</td><td> 61</td><td> 165</td>
</tr><tr>
<td> 40</td><td> 73</td><td> 181</td>
</tr> 
</table>
</td><td align=center>
<table frame=box>
<tr>
<th> Sex</th><th> State</th>
</tr>
<tr>
<td> Male</td><td> Alaska</td><td>
</tr><tr>
<td> Female</td><td> Alabama</td>
</tr><tr>
<td> Male</td><td> Alabama</td>
</tr> 
</table>
</td><td align=center>
<table frame=box>
<tr><td> </td></tr>
<tr>
<td>1</td>
</tr><tr>
<td>3.2</td>
</tr><tr>
<td>2.4</td>
</tr> 
</table>
</td></tr>
</table>
<p>In a regression, the vector would be the dependent variable, and the other columns (after factor-izing the text) the independent variables. Or think of the <a class="el" href="gentle.html#apop_data">apop_data</a> set as a partitioned matrix, where the vector is column -1, and the first column of the matrix is column zero. Here is some code to print the entire matrix, starting at column -1.</p>
<div class="fragment"><div class="line"><span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j&lt; data-&gt;matrix-&gt;size1; j++){</div>
<div class="line">    printf(<span class="stringliteral">&quot;%s\t&quot;</span>, apop_name_get(data-&gt;names, j, <span class="charliteral">&#39;r&#39;</span>));</div>
<div class="line">    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = -1; i&lt; data-&gt;matrix-&gt;size2; i++)</div>
<div class="line">        printf(<span class="stringliteral">&quot;%g\t&quot;</span>, <a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(data, j, i));</div>
<div class="line">    printf(<span class="stringliteral">&quot;\n&quot;</span>);</div>
<div class="line">}</div>
</div><!-- fragment --><p>Most functions assume that the data vector, data matrix, and text have the same row count: <code>data-&gt;vector-&gt;size==data-&gt;matrix-&gt;size1</code> and <code>data-&gt;vector-&gt;size==*data-&gt;textsize</code>. This means that the <a class="el" href="structapop__name.html">apop_name</a> structure doesn't have separate <code>vector_names</code>, <code>row_names</code>, or <code>text_row_names</code> elements: the <code>rownames</code> are assumed to apply for all.</p>
<p>See below for notes on easily managing the <code>text</code> element and the row/column names.</p>
<p>The <a class="el" href="gentle.html#apop_data">apop_data</a> set includes a <code>more</code> pointer, which will typically be <code>NULL</code>, but may point to another <a class="el" href="gentle.html#apop_data">apop_data</a> set. This is intended for a main data set and a second or third page with auxiliary information: estimated parameters on the front page and their covariance matrix on page two, or predicted data on the front page and a set of prediction intervals on page two. <code>apop_data_copy</code> and <code>apop_data_free</code> will handle all the pages of information. The <code>more</code> pointer is not intended as a linked list for millions of data points&mdash;you can probably find a way to restructure your data to use a single table (perhaps via <a class="el" href="group__conversions.html#ga7e31798ab418cbfc7488eaba27dad2af">apop_data_pack</a> and <a class="el" href="group__conversions.html#gae3058b7f39d0442f13284464d1ba235d">apop_data_unpack</a>).</p>
<p>There are a great many functions to collate, copy, merge, sort, prune, and otherwise manipulate the <a class="el" href="gentle.html#apop_data">apop_data</a> structure and its components.</p>
<ul>
<li><a class="el" href="apop_8h.html#a44b767b375fb18f61f939c6869955331">apop_data_add_named_elmt()</a> </li>
<li><a class="el" href="apop_8h.html#add67e6838ce2521a1d756894dc9b5ccf">apop_data_copy()</a> </li>
<li><a class="el" href="apop_8h.html#ad6be56d710a63aa1ceadf2242c553905">apop_data_fill()</a> </li>
<li><a class="el" href="apop_8h.html#aa24d396c7fda38bc4ab03a3ba2db4943">apop_data_memcpy()</a> </li>
<li><a class="el" href="group__conversions.html#ga7e31798ab418cbfc7488eaba27dad2af">apop_data_pack()</a> </li>
<li><a class="el" href="apop_8h.html#a1312f6568cfe832f1e35f7f44482ffb4">apop_data_rm_columns()</a> </li>
<li><a class="el" href="apop_8h.html#acefe4dee819c090e57ac69d4d8bdcfd0">apop_data_sort()</a> </li>
<li><a class="el" href="apop_8h.html#aef0284036801cb75d465a1738f65397e">apop_data_split()</a> </li>
<li><a class="el" href="apop_8h.html#a0f2ffb5c5ba2bb240e930be08fc4fd56">apop_data_stack()</a> </li>
<li><a class="el" href="apop_8h.html#a82281b8673cda3645054d7e061bb1167">apop_data_transpose()</a> </li>
<li><a class="el" href="group__conversions.html#gae3058b7f39d0442f13284464d1ba235d">apop_data_unpack()</a> </li>
<li><a class="el" href="group__convenience__fns.html#ga45bb322b9f491c46feb1e24025b6301f">apop_matrix_copy()</a> </li>
<li><a class="el" href="apop_8h.html#aa06b6ab6ec9f78f305b1dacc8ec2efd3">apop_matrix_realloc()</a> </li>
<li><a class="el" href="apop_8h.html#a5402b8254a90f2ac68770d6867502e5a">apop_matrix_rm_columns()</a> </li>
<li><a class="el" href="group__convenience__fns.html#gae50bc918e0d8b6ee913596f68ab731a8">apop_matrix_stack()</a> </li>
<li><a class="el" href="apop_8h.html#ac9c6f645acce82319fb85eb9d75112ba">apop_text_add()</a> </li>
<li><a class="el" href="apop_8h.html#a6270ad55c566140e0d7962cb4d5daacc">apop_text_paste()</a> </li>
<li><a class="el" href="group__conversions.html#ga630de5d4fcbe9ef13373029f5d813a97">apop_text_to_data()</a> </li>
<li><a class="el" href="group__convenience__fns.html#ga136ded2b6d190739deca064701cf5ca1">apop_vector_bounded()</a> </li>
<li><a class="el" href="group__convenience__fns.html#ga5baf0e8717a30fdcfba350da904c1825">apop_vector_copy()</a> </li>
<li><a class="el" href="apop_8h.html#af9b051e2fbe0e473dddeabfe4932ce0c">apop_vector_fill()</a> </li>
<li><a class="el" href="group__convenience__fns.html#gae603a7d511cf40b022f99dc69faa616a">apop_vector_stack()</a> </li>
<li><a class="el" href="apop_8h.html#af4658a3ecc0acfd17f610bc162fa5556">apop_vector_realloc()</a></li>
</ul>
<p>Apophenia builds upon the GSL, but it would be inappropriate to redundantly replicate the GSL's documentation here. You will find a link to the full GSL documentation at the end of this outline. Meanwhile, here are prototypes for a few common functions. The GSL's naming scheme is very consistent, so a simple reminder of the function name may be sufficient to indicate how they are used.</p>
<ul>
<li><code>gsl_matrix_swap_rows (gsl_matrix * m, size_t i, size_t j)</code> </li>
<li><code>gsl_matrix_swap_columns (gsl_matrix * m, size_t i, size_t j)</code> </li>
<li><code>gsl_matrix_swap_rowcol (gsl_matrix * m, size_t i, size_t j)</code> </li>
<li><code>gsl_matrix_transpose_memcpy (gsl_matrix * dest, const gsl_matrix * src)</code> </li>
<li><code>gsl_matrix_transpose (gsl_matrix * m) : square matrices only</code> </li>
<li><code>gsl_matrix_set_all (gsl_matrix * m, double x)</code> </li>
<li><code>gsl_matrix_set_zero (gsl_matrix * m)</code> </li>
<li><code>gsl_matrix_set_identity (gsl_matrix * m)</code> </li>
<li><code>void gsl_vector_set_all (gsl_vector * v, double x)</code> </li>
<li><code>void gsl_vector_set_zero (gsl_vector * v)</code> </li>
<li><code>int gsl_vector_set_basis (gsl_vector * v, size_t i)</code>: set all elements to zero, but set item <img class="formulaInl" alt="$i$" src="form_65.png"/> to one. </li>
<li><code>gsl_vector_reverse (gsl_vector * v)</code>: reverse the order of your vector's elements </li>
<li><code>gsl_vector_ptr</code> and <code>gsl_matrix_ptr</code>. To increment an element in a vector use, e.g., <code>*gsl_vector_ptr(v, 7) += 3;</code> or <code>(*gsl_vector_ptr(v, 7))++</code>.</li>
</ul>
<p><h2><a class="anchor" name="readin"><div class="trigger" onClick="showBranch('readind');swapFolder('readinf')"><img src="right.png" border="0" id="readinf" alt="pip"> Reading from text files</div></a></h2><div class="branch" id="readind">
<p>The <a class="el" href="group__conversions.html#ga630de5d4fcbe9ef13373029f5d813a97">apop_text_to_data()</a> function takes in the name of a text file with a grid of data in (comma|tab|pipe|whatever)-delimited format and reads it to a matrix. If there are names in the text file, they are copied in to the data set. See <a class="el" href="text_format.html">Notes on input text file formatting</a> for the full range and details of what can be read in.</p>
<p>If you have any columns of text, then you will need to read in via the database: use <a class="el" href="group__conversions.html#gabd00e16791a970631c3ed4ac3487da46">apop_text_to_db()</a> to convert your text file to a database table, do any database-appropriate cleaning of the input data, then use <a class="el" href="group__queries.html#gade862993166c6482f53562fb77721db9">apop_query_to_data()</a> or <a class="el" href="group__queries.html#ga1f06db34e05b5afb63112f83ad2317b8">apop_query_to_mixed_data()</a> to pull the data to an <a class="el" href="gentle.html#apop_data">apop_data</a> set.</p>
<p></div>
<p><h2><a class="anchor" name="datalloc"><div class="trigger" onClick="showBranch('datallocd');swapFolder('datallocf')"><img src="right.png" border="0" id="datallocf" alt="pip"> Alloc/free</div></a></h2><div class="branch" id="datallocd">
<ul>
<li><a class="el" href="apop_8h.html#ac39d6a56933b6f084328ec5ecd5090d8">apop_data_alloc()</a> </li>
<li><a class="el" href="apop_8h.html#a0460790f2d21f03856ed4a212a4132ec">apop_data_calloc()</a> </li>
<li><a class="el" href="apop_8h.html#a459acfde11f2c39f5c32cff377f85b9e">apop_data_free()</a> </li>
<li><a class="el" href="apop_8h.html#a9fba0f07c262a433133f6f3362617da3">apop_text_alloc()</a> : allocate or resize the text part of an <a class="el" href="gentle.html#apop_data">apop_data</a> set. </li>
<li><a class="el" href="apop_8h.html#a65386d7b37c2eaf37d1eb4e3ae7a7554">apop_text_free()</a> <pre class="fragment">See also:
</pre></li>
</ul>
<ul>
<li><code>gsl_matrix * gsl_matrix_alloc (size_t n1, size_t n2)</code> </li>
<li><code>gsl_matrix * gsl_matrix_calloc (size_t n1, size_t n2)</code> </li>
<li><code>void gsl_matrix_free (gsl_matrix * m)</code> </li>
<li><code>gsl_matrix_memcpy (gsl_matrix * dest, const gsl_matrix * src)</code> </li>
<li><code>gsl_vector * gsl_vector_alloc (size_t n)</code> </li>
<li><code>gsl_vector * gsl_vector_calloc (size_t n)</code> </li>
<li><code>void gsl_vector_free (gsl_vector * v)</code> </li>
<li><code>gsl_vector_memcpy (gsl_vector * dest, const gsl_vector * src)</code></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="gslviews"><div class="trigger" onClick="showBranch('gslviewsd');swapFolder('gslviewsf')"><img src="right.png" border="0" id="gslviewsf" alt="pip"> Using views</div></a></h2><div class="branch" id="gslviewsd">
<p>There are several macros for the common task of viewing a single row or column of a <a class="el" href="gentle.html#apop_data">apop_data</a> set.</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *d = <a class="code" href="group__queries.html#gade862993166c6482f53562fb77721db9">apop_query_to_data</a>(<span class="stringliteral">&quot;select obs1, obs2, obs3 from a_table&quot;</span>);</div>
<div class="line"></div>
<div class="line"><span class="comment">//Get a column using its name</span></div>
<div class="line"><a class="code" href="apop_8h.html#aaaf15101ec60ce66e80737375baa7391">Apop_col_t</a>(d, <span class="stringliteral">&quot;obs1&quot;</span>, ov);</div>
<div class="line"><span class="keywordtype">double</span> obs1_sum = <a class="code" href="group__convenience__fns.html#gaf6182181dd424bc4de98b4bfbd349b4e">apop_vector_sum</a>(ov);</div>
<div class="line"></div>
<div class="line"><span class="comment">//Get row zero of the data set&#39;s matrix as a vector; get its sum</span></div>
<div class="line"><span class="keywordtype">double</span> first_row_sum = <a class="code" href="group__convenience__fns.html#gaf6182181dd424bc4de98b4bfbd349b4e">apop_vector_sum</a>(<a class="code" href="apop_8h.html#a314f28de914dfe6e45170e09d71d5429">Apop_rv</a>(d, 0));</div>
<div class="line"></div>
<div class="line"><span class="comment">//Get a row or rows as a standalone one-row apop_data set</span></div>
<div class="line"><a class="code" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print</a>(<a class="code" href="apop_8h.html#a401dbc60c9915be8e4c4def68b22d755">Apop_r</a>(d, 0));</div>
<div class="line"></div>
<div class="line"><span class="comment">//ten rows starting at row 3:</span></div>
<div class="line">Apop_data_rows(d, 3, 10, d10);</div>
<div class="line"><a class="code" href="group__output.html#ga9a30c437ec11a11e37e0db7f82386275">apop_data_show</a>(d10);</div>
<div class="line"></div>
<div class="line"><span class="comment">//First column&#39;s sum</span></div>
<div class="line"><span class="keywordtype">double</span> first_col_sum = <a class="code" href="group__convenience__fns.html#gaf6182181dd424bc4de98b4bfbd349b4e">apop_vector_sum</a>(<a class="code" href="apop_8h.html#a825321182fbaa7471069f13447170810">Apop_cv</a>(d, 0));</div>
<div class="line"></div>
<div class="line"><span class="comment">//Pull a 10x5 submatrix, whose first element is the (2,3)rd</span></div>
<div class="line"><span class="comment">//element of the parent data set&#39;s matrix</span></div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> sub_sum = <a class="code" href="group__convenience__fns.html#ga369c774aac917e56d8e32f9c66d9c2a0">apop_matrix_sum</a>(<a class="code" href="apop_8h.html#ad0375e20079df9652337e39c1317a14d">Apop_subm</a>(d, 2,3, 10,5));</div>
</div><!-- fragment --><p>To make it easier to use the result of these macros as an argument to a function, these macros have abbreviated names.</p>
<ul>
<li><a class="el" href="apop_8h.html#adabbf2a1717d5c76a834e090ddfcfeaa">Apop_c</a> </li>
<li><a class="el" href="apop_8h.html#a401dbc60c9915be8e4c4def68b22d755">Apop_r</a> </li>
<li><a class="el" href="apop_8h.html#a825321182fbaa7471069f13447170810">Apop_cv</a> </li>
<li><a class="el" href="apop_8h.html#a314f28de914dfe6e45170e09d71d5429">Apop_rv</a> </li>
<li><a class="el" href="apop_8h.html#afc073d46416446dc4f0cdb9e4f1b9a03">Apop_cs</a> </li>
<li><a class="el" href="apop_8h.html#a585f5cf3125fb5fc875dea85980eabd1">Apop_rs</a> </li>
<li><a class="el" href="apop_8h.html#ad0375e20079df9652337e39c1317a14d">Apop_subm</a></li>
</ul>
<p>A second set of macros have a slightly different syntax, taking the name of the object to be declared as the last argument. These can not be used as expressions such as function arguments.</p>
<ul>
<li><a class="el" href="apop_8h.html#aaaf15101ec60ce66e80737375baa7391">Apop_col_t</a> </li>
<li><a class="el" href="apop_8h.html#a7d87af4e65acbf6982a34cf3dc33eae6">Apop_row_t</a> </li>
<li><a class="el" href="apop_8h.html#ad56ce572c7489820777290153ab7d87b">Apop_col_tv</a> </li>
<li><a class="el" href="apop_8h.html#a6396c26f1d47cced2be3a761185d3b35">Apop_row_tv</a> </li>
<li><a class="el" href="apop_8h.html#a9abee86763f92d505fd4fb792f06fd7c">Apop_matrix_row</a> </li>
<li><a class="el" href="apop_8h.html#a3e7f99e26eb4d5d5438fd5fa16fcadde">Apop_matrix_col</a></li>
</ul>
<p>The view is an automatic variable, not a pointer, and therefore disappears at the end of the scope in which it is declared. If you want to retain the data after the function exits, copy it to another vector:</p>
<div class="fragment"><div class="line"><a class="code" href="apop_8h.html#a9abee86763f92d505fd4fb792f06fd7c">Apop_matrix_row</a>(d-&gt;matrix, 2, rowtwo);</div>
<div class="line">gsl_vector *outvector = <a class="code" href="group__convenience__fns.html#ga5baf0e8717a30fdcfba350da904c1825">apop_vector_copy</a>(rowtwo);</div>
</div><!-- fragment --><p>Curly braces always delimit scope, not just at the end of a function. These macros work by generating a number of local variables, which you may be able to see in your debugger. When program evaluation exits a given block, all variables in that block are erased. Here is some sample code that won't work:</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *outdata;</div>
<div class="line"><span class="keywordflow">if</span> (get_odd){</div>
<div class="line">    outdata = <a class="code" href="apop_8h.html#a401dbc60c9915be8e4c4def68b22d755">Apop_r</a>(data, 1);</div>
<div class="line">} <span class="keywordflow">else</span> {</div>
<div class="line">    outdata = <a class="code" href="apop_8h.html#a401dbc60c9915be8e4c4def68b22d755">Apop_r</a>(data, 0);</div>
<div class="line">}</div>
<div class="line"><a class="code" href="group__output.html#ga9a30c437ec11a11e37e0db7f82386275">apop_data_show</a>(outdata); <span class="comment">//breaks: outdata points to out-of-scope variables.</span></div>
</div><!-- fragment --><p>For this if/then statement, there are two sets of local variables generated: one for the <code>if</code> block, and one for the <code>else</code> block. By the last line, neither exists. You can get around the problem here by making sure to not put the macro declaring new variables in a block. E.g.:</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *outdata = <a class="code" href="apop_8h.html#a401dbc60c9915be8e4c4def68b22d755">Apop_r</a>(data, get_odd ? 1 : 0);</div>
<div class="line"><a class="code" href="group__output.html#ga9a30c437ec11a11e37e0db7f82386275">apop_data_show</a>(outdata);</div>
</div><!-- fragment --><p>This is a general rule about how variables declared in blocks will behave, but because the macros obscure the variable declarations, it is especially worth watching out for here.</p>
<p></div>
<p><h2><a class="anchor" name="setgetsec"><div class="trigger" onClick="showBranch('setgetsecd');swapFolder('setgetsecf')"><img src="right.png" border="0" id="setgetsecf" alt="pip"> Set/get</div></a></h2><div class="branch" id="setgetsecd">
<p>The set/get functions can act on both names or indices. Sample usages:</p>
<div class="fragment"><div class="line"><span class="keywordtype">double</span> twothree = <a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(data, 2, 3); <span class="comment">//just indices</span></div>
<div class="line"><a class="code" href="group__data__set__get.html#gaea55629952c595bc8211d3333077dc51">apop_data_set</a>(data, .rowname=<span class="stringliteral">&quot;A row&quot;</span>, .colname=<span class="stringliteral">&quot;this column&quot;</span>, .val=13);</div>
<div class="line"><span class="keywordtype">double</span> AIC = <a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(your_model-&gt;info, .rowname=<span class="stringliteral">&quot;AIC&quot;</span>);</div>
</div><!-- fragment --><ul>
<li><a class="el" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get()</a> </li>
<li><a class="el" href="group__data__set__get.html#gaea55629952c595bc8211d3333077dc51">apop_data_set()</a> </li>
<li><a class="el" href="group__data__set__get.html#ga59d1c0299c7eb87224bc6c60555784fe">apop_data_ptr()</a> : returns a pointer to the element. <pre class="fragment">See also:
</pre></li>
</ul>
<ul>
<li><code>double gsl_matrix_get (const gsl_matrix * m, size_t i, size_t j)</code> </li>
<li><code>double gsl_vector_get (const gsl_vector * v, size_t i)</code> </li>
<li><code>void gsl_matrix_set (gsl_matrix * m, size_t i, size_t j, double x)</code> </li>
<li><code>void gsl_vector_set (gsl_vector * v, size_t i, double x)</code> </li>
<li><code>double * gsl_matrix_ptr (gsl_matrix * m, size_t i, size_t j)</code> </li>
<li><code>double * gsl_vector_ptr (gsl_vector * v, size_t i)</code> </li>
<li><code>const double * gsl_matrix_const_ptr (const gsl_matrix * m, size_t i, size_t j)</code> </li>
<li><code>const double * gsl_vector_const_ptr (const gsl_vector * v, size_t i)</code> </li>
<li><code>gsl_matrix_get_row (gsl_vector * v, const gsl_matrix * m, size_t i)</code> </li>
<li><code>gsl_matrix_get_col (gsl_vector * v, const gsl_matrix * m, size_t j)</code> </li>
<li><code>gsl_matrix_set_row (gsl_matrix * m, size_t i, const gsl_vector * v)</code> </li>
<li><code>gsl_matrix_set_col (gsl_matrix * m, size_t j, const gsl_vector * v)</code></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="mapplysec"><div class="trigger" onClick="showBranch('mapplysecd');swapFolder('mapplysecf')"><img src="right.png" border="0" id="mapplysecf" alt="pip"> Map/apply</div></a></h2><div class="branch" id="mapplysecd">
<p><a class="anchor" id="outline_mapply"></a>These functions allow you to send each element of a vector or matrix to a function, either producing a new matrix (map) or transforming the original (apply). The <code></code>..._sum functions return the sum of the mapped output.</p>
<p>There is an older and a newer set of functions. The older versions, which act on <code>gsl_matrix</code>es or <code>gsl_vector</code>s have more verbose names; the newer versions, which take in an <a class="el" href="gentle.html#apop_data">apop_data</a> set, use the <a class="el" href="designated.html">Designated initializers</a> syntax to add a few options and a more brief syntax.</p>
<p>You can do many things quickly with these functions.</p>
<p>Get the sum of squares of a vector's elements:</p>
<div class="fragment"><div class="line">  <span class="comment">//given apop_data *dataset and gsl_vector *v:</span></div>
<div class="line"><span class="keywordtype">double</span> sum_of_squares = <a class="code" href="group__mapply.html#ga30601e357a1e21648f8d8db3c9ed6f3a">apop_map_sum</a>(dataset, gsl_pow_2);</div>
<div class="line"><span class="keywordtype">double</span> sum_of_squares = <a class="code" href="group__mapply.html#gaa1a135fcdd8ea6f9cbc0d518773d6c3c">apop_vector_map_sum</a>(v, gsl_pow_2);</div>
</div><!-- fragment --><p>Here, we create an index vector [ <img class="formulaInl" alt="$0, 1, 2, ...$" src="form_102.png"/>].</p>
<div class="fragment"><div class="line"><span class="keywordtype">double</span> index(<span class="keywordtype">double</span> in, <span class="keywordtype">int</span> index){<span class="keywordflow">return</span> index;}</div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *d = <a class="code" href="apop_8h.html#ac39d6a56933b6f084328ec5ecd5090d8">apop_data_alloc</a>(100);</div>
<div class="line"><a class="code" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a>(d, .fn_di=index, .inplace=<span class="charliteral">&#39;y&#39;</span>);</div>
</div><!-- fragment --><p>Given your log likelihood function and a data set where each row of the matrix is an observation, find the total log likelihood:</p>
<div class="fragment"><div class="line"><span class="keyword">static</span> <span class="keywordtype">double</span> your_log_likelihood_fn(<span class="keyword">const</span> gsl_vector * in)</div>
<div class="line">     {[your math goes here]}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> total_ll = <a class="code" href="group__mapply.html#ga4f0c51a046140b563c8e12a6032e9a16">apop_matrix_map_sum</a>(dataset, your_log_likelihood_fn);</div>
</div><!-- fragment --><p>How many missing elements are there in your data matrix?</p>
<div class="fragment"><div class="line"><span class="keyword">static</span> <span class="keywordtype">double</span> nan_check(<span class="keyword">const</span> <span class="keywordtype">double</span> in){ <span class="keywordflow">return</span> isnan(in);}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">int</span> missing_ct = <a class="code" href="group__mapply.html#ga30601e357a1e21648f8d8db3c9ed6f3a">apop_map_sum</a>(in, nan_check, .part=<span class="charliteral">&#39;m&#39;</span>);</div>
</div><!-- fragment --><p>Get the mean of the not-NaN elements of a data set:</p>
<div class="fragment"><div class="line"><span class="keyword">static</span> <span class="keywordtype">double</span> no_nan_val(<span class="keyword">const</span> <span class="keywordtype">double</span> in){ <span class="keywordflow">return</span> isnan(in)? 0 : in;}</div>
<div class="line"><span class="keyword">static</span> <span class="keywordtype">double</span> not_nan_check(<span class="keyword">const</span> <span class="keywordtype">double</span> in){ <span class="keywordflow">return</span> !isnan(in);}</div>
<div class="line"></div>
<div class="line"><span class="keyword">static</span> <span class="keywordtype">double</span> apop_mean_no_nans(<a class="code" href="structapop__data.html">apop_data</a> *in){</div>
<div class="line">    <span class="keywordflow">return</span> <a class="code" href="group__mapply.html#ga30601e357a1e21648f8d8db3c9ed6f3a">apop_map_sum</a>(in, no_nan_val)/<a class="code" href="group__mapply.html#ga30601e357a1e21648f8d8db3c9ed6f3a">apop_map_sum</a>(in, not_nan_check);</div>
<div class="line">}</div>
</div><!-- fragment --><p>The following program randomly generates a data set where each row is a list of numbers with a different mean. It then finds the <img class="formulaInl" alt="$t$" src="form_44.png"/> statistic for each row, and the confidence with which we reject the claim that the statistic is less than or equal to zero.</p>
<p>Notice how the older <a class="el" href="group__mapply.html#ga51964750c64d327e52e2230019e44f66">apop_vector_apply</a> uses file-global variables to pass information into the functions, while the <a class="el" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a> uses a pointer to the constant parameters to input to the functions.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="apop_8h.html">apop.h</a>&gt;</span></div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> row_offset;</div>
<div class="line">gsl_rng *r;</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">void</span> offset_rng(<span class="keywordtype">double</span> *v){*v = gsl_rng_uniform(r) + row_offset;}</div>
<div class="line"><span class="keywordtype">double</span> find_tstat(gsl_vector *in){ <span class="keywordflow">return</span> <a class="code" href="group__vector__moments.html#ga0f69c8414f1d93fa15f3a631710dcd7a">apop_mean</a>(in)/sqrt(<a class="code" href="group__vector__moments.html#gacc311abe133963c7959162a7d3d83fb8">apop_var</a>(in));}</div>
<div class="line"><span class="keywordtype">double</span> conf(<span class="keywordtype">double</span> in, <span class="keywordtype">void</span> *df){ <span class="keywordflow">return</span> gsl_cdf_tdist_P(in, *(<span class="keywordtype">int</span> *)df);}</div>
<div class="line"></div>
<div class="line"><span class="comment">//apop_vector_mean is a macro, so we can&#39;t point a pointer to it.</span></div>
<div class="line"><span class="keywordtype">double</span> mu(gsl_vector *in){ <span class="keywordflow">return</span> <a class="code" href="apop_8h.html#ab20e5ad95df0f24a1ea06294695031a8">apop_vector_mean</a>(in);}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">int</span> main(){</div>
<div class="line">    <a class="code" href="structapop__data.html">apop_data</a> *d = <a class="code" href="apop_8h.html#ac39d6a56933b6f084328ec5ecd5090d8">apop_data_alloc</a>(10, 100);</div>
<div class="line">    r = <a class="code" href="group__convenience__fns.html#gada044cf02135422a9f9292b6054c86dd">apop_rng_alloc</a>(3242);</div>
<div class="line">    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt; 10; i++){</div>
<div class="line">        row_offset = gsl_rng_uniform(r)*2 -1;</div>
<div class="line">        Apop_row_v(d, i, onerow);</div>
<div class="line">        <a class="code" href="group__mapply.html#ga51964750c64d327e52e2230019e44f66">apop_vector_apply</a>(onerow, offset_rng);</div>
<div class="line">    }</div>
<div class="line"></div>
<div class="line">    <span class="keywordtype">size_t</span> df = d-&gt;matrix-&gt;size2-1;</div>
<div class="line">    <a class="code" href="structapop__data.html">apop_data</a> *means = <a class="code" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a>(d, .fn_v = mu, .part =<span class="charliteral">&#39;r&#39;</span>);</div>
<div class="line">    <a class="code" href="structapop__data.html">apop_data</a> *tstats = <a class="code" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a>(d, .fn_v = find_tstat, .part =<span class="charliteral">&#39;r&#39;</span>);</div>
<div class="line">    <a class="code" href="structapop__data.html">apop_data</a> *confidences = <a class="code" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a>(tstats, .fn_dp = conf, .param = &amp;df);</div>
<div class="line"></div>
<div class="line"><span class="preprocessor">#ifndef Testing</span></div>
<div class="line">    printf(<span class="stringliteral">&quot;means:\n&quot;</span>); <a class="code" href="group__output.html#ga9a30c437ec11a11e37e0db7f82386275">apop_data_show</a>(means);</div>
<div class="line">    printf(<span class="stringliteral">&quot;\nt stats:\n&quot;</span>); <a class="code" href="group__output.html#ga9a30c437ec11a11e37e0db7f82386275">apop_data_show</a>(tstats);</div>
<div class="line">    printf(<span class="stringliteral">&quot;\nconfidences:\n&quot;</span>); <a class="code" href="group__output.html#ga9a30c437ec11a11e37e0db7f82386275">apop_data_show</a>(confidences);</div>
<div class="line"><span class="preprocessor">#endif</span></div>
<div class="line"></div>
<div class="line">    <span class="comment">//Some sanity checks, for Apophenia&#39;s test suite.</span></div>
<div class="line">    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt; 10; i++){</div>
<div class="line">        <span class="comment">//sign of mean == sign of t stat.</span></div>
<div class="line">        assert(<a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(means, i, -1) * <a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(tstats, i, -1) &gt;=0);</div>
<div class="line"></div>
<div class="line">        <span class="comment">//inverse of P-value should be the t statistic.</span></div>
<div class="line">        assert(fabs(gsl_cdf_tdist_Pinv(<a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(confidences, i, -1),100) </div>
<div class="line">                    - <a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(tstats, i, -1) &lt; 1e-3));</div>
<div class="line">    }</div>
<div class="line">}</div>
</div><!-- fragment --><p>One more toy example, demonstrating the use of <a class="el" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a> and <a class="el" href="group__mapply.html#ga30601e357a1e21648f8d8db3c9ed6f3a">apop_map_sum</a> :</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="apop_8h.html">apop.h</a>&gt;</span></div>
<div class="line"><span class="comment">/* This sample code sets the elements of a data set&#39;s vector to one</span></div>
<div class="line"><span class="comment">   if the index is even.  Then, via the weights vector, it adds up</span></div>
<div class="line"><span class="comment">   the even indices.</span></div>
<div class="line"><span class="comment"></span></div>
<div class="line"><span class="comment">   There is really no need to use the weights vector; this code</span></div>
<div class="line"><span class="comment">   snippet is an element of Apophenia&#39;s test suite, and goes the long</span></div>
<div class="line"><span class="comment">   way to test that the weights are correctly handled. */</span></div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> set_vector_to_even(<a class="code" href="structapop__data.html">apop_data</a> * r, <span class="keywordtype">int</span> index){</div>
<div class="line">    <a class="code" href="group__data__set__get.html#gaea55629952c595bc8211d3333077dc51">apop_data_set</a>(r, 0, -1, .val=1-(index %2));</div>
<div class="line">    <span class="keywordflow">return</span> 0;</div>
<div class="line">}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> set_weight_to_index(<a class="code" href="structapop__data.html">apop_data</a> * r, <span class="keywordtype">int</span> index){ </div>
<div class="line">    gsl_vector_set(r-&gt;weights, 0, index); </div>
<div class="line">    <span class="keywordflow">return</span> 0;</div>
<div class="line">}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> weight_given_even(<a class="code" href="structapop__data.html">apop_data</a> *r){ </div>
<div class="line">    <span class="keywordflow">return</span> gsl_vector_get(r-&gt;vector, 0) ? gsl_vector_get(r-&gt;weights, 0) : 0; </div>
<div class="line">}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">int</span> main(){</div>
<div class="line">    <a class="code" href="structapop__data.html">apop_data</a> *d = <a class="code" href="apop_8h.html#ac39d6a56933b6f084328ec5ecd5090d8">apop_data_alloc</a>(100);</div>
<div class="line">    d-&gt;weights = gsl_vector_alloc(100);</div>
<div class="line">    <a class="code" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a>(d, .fn_ri=set_vector_to_even, .inplace=<span class="charliteral">&#39;v&#39;</span>);</div>
<div class="line">    <a class="code" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map</a>(d, .fn_ri=set_weight_to_index, .inplace=<span class="charliteral">&#39;v&#39;</span>);</div>
<div class="line">    <span class="keywordtype">double</span> sum = <a class="code" href="group__mapply.html#ga30601e357a1e21648f8d8db3c9ed6f3a">apop_map_sum</a>(d, .fn_r = weight_given_even);</div>
<div class="line">    assert(sum == 49*25*2);</div>
<div class="line">}</div>
</div><!-- fragment --><ul>
<li><a class="el" href="group__mapply.html#gaacff3707e418a067104c77c410b43f69">apop_map()</a> </li>
<li><a class="el" href="group__mapply.html#ga30601e357a1e21648f8d8db3c9ed6f3a">apop_map_sum()</a> </li>
<li><a class="el" href="group__mapply.html#ga5c4ae58b861199bbc1a460719942d98f">apop_matrix_apply()</a> </li>
<li><a class="el" href="group__mapply.html#ga5da5091a11c37d3501d678b8ecec466f">apop_matrix_map()</a> </li>
<li><a class="el" href="group__mapply.html#gae04bc576579bfb2e14d5004d4af7bb88">apop_matrix_map_all_sum()</a> </li>
<li><a class="el" href="group__mapply.html#ga4f0c51a046140b563c8e12a6032e9a16">apop_matrix_map_sum()</a> </li>
<li><a class="el" href="group__mapply.html#ga51964750c64d327e52e2230019e44f66">apop_vector_apply()</a> </li>
<li><a class="el" href="group__mapply.html#gaf30eadf8772ecb9a2689ddf104902ea4">apop_vector_map()</a> </li>
<li><a class="el" href="group__mapply.html#gaa1a135fcdd8ea6f9cbc0d518773d6c3c">apop_vector_map_sum()</a></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="matrixmathtwo"><div class="trigger" onClick="showBranch('matrixmathtwod');swapFolder('matrixmathtwof')"><img src="right.png" border="0" id="matrixmathtwof" alt="pip"> Basic Math</div></a></h2><div class="branch" id="matrixmathtwod">
<ul>
<li><a class="el" href="group__convenience__fns.html#ga032dcecb67155c527c884dd2c1cc6a6e">apop_vector_exp</a> : exponentiate every element of a vector </li>
<li><a class="el" href="group__convenience__fns.html#ga5a622b8360eba63795a6049c30ac54a2">apop_vector_log</a> : take the log of every element of a vector </li>
<li><a class="el" href="group__convenience__fns.html#ga1a4a67e0b0bec8205cfe75bd3786c30e">apop_vector_log10</a> : take the log (base 10) of every element of a vector </li>
<li><a class="el" href="group__convenience__fns.html#ga44a77f5e6867c627fd313f3f6e252ff8">apop_vector_distance</a> : find the distance between two vectors via various metrics </li>
<li><a class="el" href="apop__stats_8c.html#a194ad92b1bc632ba1d13d9958c7b5b4d">apop_vector_normalize</a> : scale/shift a matrix to have mean zero, sum to one, et cetera </li>
<li><a class="el" href="apop__stats_8c.html#af6290c453c3012c1d0a6987688ebe795">apop_matrix_normalize</a> : apply <a class="el" href="apop__stats_8c.html#a194ad92b1bc632ba1d13d9958c7b5b4d">apop_vector_normalize</a> to every column or row of a matrix </li>
<li>apop_vector_entropy: calculate the entropy of a vector of frequencies or probabilities <pre class="fragment">See also:
</pre></li>
</ul>
<ul>
<li><code>int gsl_matrix_add (gsl_matrix * a, const gsl_matrix * b)</code> </li>
<li><code>int gsl_matrix_sub (gsl_matrix * a, const gsl_matrix * b)</code> </li>
<li><code>int gsl_matrix_mul_elements (gsl_matrix * a, const gsl_matrix * b)</code> </li>
<li><code>int gsl_matrix_div_elements (gsl_matrix * a, const gsl_matrix * b)</code> </li>
<li><code>int gsl_matrix_scale (gsl_matrix * a, const double x)</code> </li>
<li><code>int gsl_matrix_add_constant (gsl_matrix * a, const double x)</code> </li>
<li><code>gsl_vector_add (gsl_vector * a, const gsl_vector * b)</code> </li>
<li><code>gsl_vector_sub (gsl_vector * a, const gsl_vector * b)</code> </li>
<li><code>gsl_vector_mul (gsl_vector * a, const gsl_vector * b)</code> </li>
<li><code>gsl_vector_div (gsl_vector * a, const gsl_vector * b)</code> </li>
<li><code>gsl_vector_scale (gsl_vector * a, const double x)</code> </li>
<li><code>gsl_vector_add_constant (gsl_vector * a, const double x)</code></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="matrixmath"><div class="trigger" onClick="showBranch('matrixmathd');swapFolder('matrixmathf')"><img src="right.png" border="0" id="matrixmathf" alt="pip"> Matrix math</div></a></h2><div class="branch" id="matrixmathd">
<ul>
<li><a class="el" href="group__linear__algebra.html#gaf2d13e1b1f20f97e0772a0be41dc430e">apop_dot()</a>: matrix <img class="formulaInl" alt="$\cdot$" src="form_33.png"/> matrix, matrix <img class="formulaInl" alt="$\cdot$" src="form_33.png"/> vector, or vector <img class="formulaInl" alt="$\cdot$" src="form_33.png"/> matrix </li>
<li><a class="el" href="group__linear__algebra.html#ga03038130e4ba55228f9e164dba513961">apop_matrix_determinant</a> </li>
<li><a class="el" href="group__linear__algebra.html#ga185243be2d8a93686c68263cc3318f47">apop_matrix_inverse</a> </li>
<li><a class="el" href="group__linear__algebra.html#ga6984006ab44fa7c9c2e99463f7d4081b">apop_det_and_inv()</a>: find determinant and inverse at the same time</li>
</ul>
<p>See the GSL documentation for voluminous further options.</p>
<p></div>
<p><h2><a class="anchor" name="sumstats"><div class="trigger" onClick="showBranch('sumstatsd');swapFolder('sumstatsf')"><img src="right.png" border="0" id="sumstatsf" alt="pip"> Summary stats</div></a></h2><div class="branch" id="sumstatsd">
<ul>
<li><a class="el" href="group__output.html#gaca60728fc3b691acfcadebf3a8535532">apop_data_summarize</a> () </li>
<li><a class="el" href="apop_8h.html#acfa62ab9157d4d92338a29093c7d1b30">apop_vector_moving_average()</a> </li>
<li><a class="el" href="apop_8h.html#afb838084f967543fa8a03e460180d53b">apop_vector_percentiles()</a> <pre class="fragment">    See also:
</pre></li>
</ul>
<ul>
<li><code>double gsl_matrix_max (const gsl_matrix * m)</code> </li>
<li><code>double gsl_matrix_min (const gsl_matrix * m)</code> </li>
<li><code>void gsl_matrix_minmax (const gsl_matrix * m, double * min_out, double * max_out)</code> </li>
<li><code>void gsl_matrix_max_index (const gsl_matrix * m, size_t * imax, size_t * jmax)</code> </li>
<li><code>void gsl_matrix_min_index (const gsl_matrix * m, size_t * imin, size_t * jmin)</code> </li>
<li><code>void gsl_matrix_minmax_index (const gsl_matrix * m, size_t * imin, size_t * jmin, size_t * imax, size_t * jmax)</code> </li>
<li><code>gsl_vector_max (const gsl_vector * v)</code> </li>
<li><code>gsl_vector_min (const gsl_vector * v)</code> </li>
<li><code>gsl_vector_minmax (const gsl_vector * v, double * min_out, double * max_out)</code> </li>
<li><code>gsl_vector_max_index (const gsl_vector * v)</code> </li>
<li><code>gsl_vector_min_index (const gsl_vector * v)</code> </li>
<li><code>gsl_vector_minmax_index (const gsl_vector * v, size_t * imin, size_t * imax)</code></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="moments"><div class="trigger" onClick="showBranch('momentsd');swapFolder('momentsf')"><img src="right.png" border="0" id="momentsf" alt="pip"> Moments</div></a></h2><div class="branch" id="momentsd">
<p>For most of these, you can add a weights vector for weighted mean/var/cov/..., such as <code>apop_vector_mean(d-&gt;vector, .weights=d-&gt;weights)</code></p>
<ul>
<li><a class="el" href="group__vector__moments.html#ga0f69c8414f1d93fa15f3a631710dcd7a">apop_mean()</a>: the first three with short names operate on a vector. </li>
<li><a class="el" href="group__convenience__fns.html#ga7f85a98b61078ba3dc273f17eddd34a7">apop_sum()</a> </li>
<li><a class="el" href="group__vector__moments.html#gacc311abe133963c7959162a7d3d83fb8">apop_var()</a> </li>
<li><a class="el" href="group__convenience__fns.html#ga369c774aac917e56d8e32f9c66d9c2a0">apop_matrix_sum</a> () </li>
<li><a class="el" href="apop__stats_8c.html#afa3ab967f57f8ad3cfb99a4a82cfae02">apop_data_correlation</a> () </li>
<li><a class="el" href="apop__stats_8c.html#aeae37a5bf11cddee59e272b083a4efc7">apop_data_covariance</a> () </li>
<li><a class="el" href="group__output.html#gaca60728fc3b691acfcadebf3a8535532">apop_data_summarize</a> () </li>
<li><a class="el" href="group__convenience__fns.html#ga040b9e9fea1abe105de86081bf396d40">apop_matrix_mean</a> () </li>
<li><a class="el" href="group__convenience__fns.html#gaf794cc2491ef147e0c67eb05feed7a87">apop_matrix_mean_and_var</a> () </li>
<li><a class="el" href="group__vector__moments.html#gae1154897fec0c56bc0161ed24c48759c">apop_vector_correlation</a> () </li>
<li><a class="el" href="apop__stats_8c.html#aeb71737b8996eb15f7698246676a78dd">apop_vector_cov</a> () </li>
<li><a class="el" href="group__vector__moments.html#gaf5740c698174b1a265d182f4af1c0252">apop_vector_kurtosis</a> () </li>
<li><a class="el" href="group__vector__moments.html#gaabe3183e8a6a496f3fd6ab316687381e">apop_vector_kurtosis_pop</a> () </li>
<li><a class="el" href="apop_8h.html#ab20e5ad95df0f24a1ea06294695031a8">apop_vector_mean()</a> </li>
<li><a class="el" href="group__vector__moments.html#gac9d269f1dc62b038c0da72085cac1c28">apop_vector_skew()</a> </li>
<li><a class="el" href="group__vector__moments.html#gafb284d668d567de4810620962043f827">apop_vector_skew_pop()</a> </li>
<li><a class="el" href="group__convenience__fns.html#gaf6182181dd424bc4de98b4bfbd349b4e">apop_vector_sum()</a> </li>
<li><a class="el" href="apop_8h.html#adcc397ab90933d79b3062206301a0aa0">apop_vector_var()</a> </li>
<li><a class="el" href="group__vector__moments.html#ga2446385225c68c62ade26402c63600a4">apop_vector_var_m</a> ()</li>
</ul>
<p></div>
<p><h2><a class="anchor" name="convsec"><div class="trigger" onClick="showBranch('convsecd');swapFolder('convsecf')"><img src="right.png" border="0" id="convsecf" alt="pip"> Conversion among types</div></a></h2><div class="branch" id="convsecd">
<p>There are no functions provided to convert from <a class="el" href="gentle.html#apop_data">apop_data</a> to the constituent elements, because you don't need a function.</p>
<p>If you need an individual element, you can just use its pointer to retrieve it:</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *d = apop_text_to_mixed_data(<span class="stringliteral">&quot;vmw&quot;</span>, <span class="stringliteral">&quot;select result, age, &quot;</span></div>
<div class="line">                                     <span class="stringliteral">&quot;income, sampleweight from data&quot;</span>);</div>
<div class="line"><span class="keywordtype">double</span> avg_result = <a class="code" href="apop_8h.html#ab20e5ad95df0f24a1ea06294695031a8">apop_vector_mean</a>(d-&gt;vector, .weights=d-&gt;weights);</div>
</div><!-- fragment --><p>In the other direction, you can use compound literals to wrap an <a class="el" href="gentle.html#apop_data">apop_data</a> struct around a loose vector or matrix:</p>
<div class="fragment"><div class="line"><span class="comment">//Given:</span></div>
<div class="line">gsl_vector *v;</div>
<div class="line">gsl_matrix *m;</div>
<div class="line"></div>
<div class="line"><span class="comment">// Then this form wraps the elements into \ref apop_data structs. Note that</span></div>
<div class="line"><span class="comment">// these are not pointers: they&#39;re automatically allocated and therefore</span></div>
<div class="line"><span class="comment">// the extra memory use for the wrapper is cleaned up on exit from scope.</span></div>
<div class="line"></div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *dv = &amp;(<a class="code" href="apop_8h.html#a0c8c2a99fb8c6ec54822aca08591545e">apop_data</a>){.vector=v}; </div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *dm = &amp;(<a class="code" href="apop_8h.html#a0c8c2a99fb8c6ec54822aca08591545e">apop_data</a>){.matrix=m};</div>
<div class="line"></div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *v_dot_m = <a class="code" href="group__linear__algebra.html#gaf2d13e1b1f20f97e0772a0be41dc430e">apop_dot</a>(dv, dm);</div>
<div class="line"></div>
<div class="line"><span class="comment">//Here is a macro to hide C&#39;s ugliness:</span></div>
<div class="line"><span class="preprocessor">#define As_data(...) (&amp;(apop_data){__VA_ARGS__})</span></div>
<div class="line"></div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *v_dot_m2 = <a class="code" href="group__linear__algebra.html#gaf2d13e1b1f20f97e0772a0be41dc430e">apop_dot</a>(As_data(.vector=v), As_data(.matrix=m));</div>
<div class="line"></div>
<div class="line"><span class="comment">//The wrapped object is an automatically-allocated structure pointing to the</span></div>
<div class="line"><span class="comment">//original data. If it needs to persist or be separate from the original,</span></div>
<div class="line"><span class="comment">//make a copy:</span></div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *dm_copy = <a class="code" href="apop_8h.html#add67e6838ce2521a1d756894dc9b5ccf">apop_data_copy</a>(As_data(.vector=v, .matrix=m));</div>
</div><!-- fragment --><ul>
<li><a class="el" href="group__conversions.html#ga8abdcdbaceaaf0dd3882a164bc2565de">apop_array_to_vector()</a> : <code>double*</code> <img class="formulaInl" alt="$\to$" src="form_41.png"/> <code>gsl_vector</code> </li>
<li><a class="el" href="apop_8h.html#ad6be56d710a63aa1ceadf2242c553905">apop_data_fill()</a> : <code>double*</code> <img class="formulaInl" alt="$\to$" src="form_41.png"/> <a class="el" href="gentle.html#apop_data">apop_data</a>. See also <a class="el" href="apop_8h.html#acb6b2c53478f9db902ba0340ba499819">apop_data_falloc</a> </li>
<li><a class="el" href="group__conversions.html#ga630de5d4fcbe9ef13373029f5d813a97">apop_text_to_data()</a> : delimited text file <img class="formulaInl" alt="$\to$" src="form_41.png"/> <a class="el" href="gentle.html#apop_data">apop_data</a> </li>
<li><a class="el" href="group__conversions.html#gabd00e16791a970631c3ed4ac3487da46">apop_text_to_db()</a> : delimited text file <img class="formulaInl" alt="$\to$" src="form_41.png"/> database </li>
<li><a class="el" href="group__conversions.html#ga604388dc6d7d08d95bbaa4422085db9c">apop_vector_to_matrix()</a></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="names"><div class="trigger" onClick="showBranch('namesd');swapFolder('namesf')"><img src="right.png" border="0" id="namesf" alt="pip"> Name handling</div></a></h2><div class="branch" id="namesd">
<p>If you generate your data set from the database via <a class="el" href="group__queries.html#gade862993166c6482f53562fb77721db9">apop_query_to_data</a> (or <a class="el" href="group__queries.html#ga2fbd580cd71f481ae04ee03b5cf64ff5">apop_query_to_text</a> or <a class="el" href="group__queries.html#ga1f06db34e05b5afb63112f83ad2317b8">apop_query_to_mixed_data</a>) then column names appear as expected. Set <code>apop_opts.db_name_column</code> to the name of a column in your query result to use that column name for row names.</p>
<p>Sample uses, given <a class="el" href="gentle.html#apop_data">apop_data</a> set <code>d</code>:</p>
<div class="fragment"><div class="line"><span class="keywordtype">int</span> row_name_count = d-&gt;names-&gt;rowct</div>
<div class="line"><span class="keywordtype">int</span> col_name_count = d-&gt;names-&gt;colct</div>
<div class="line"><span class="keywordtype">int</span> text_name_count = d-&gt;names-&gt;textct</div>
<div class="line"></div>
<div class="line"><span class="comment">//Manually add a name:</span></div>
<div class="line"><a class="code" href="apop_8h.html#ae32d4f48cff7788b617d40155703ae64">apop_name_add</a>(d-&gt;names, <span class="stringliteral">&quot;the vector&quot;</span>, <span class="charliteral">&#39;v&#39;</span>);</div>
<div class="line"><a class="code" href="apop_8h.html#ae32d4f48cff7788b617d40155703ae64">apop_name_add</a>(d-&gt;names, <span class="stringliteral">&quot;row 0&quot;</span>, <span class="charliteral">&#39;r&#39;</span>);</div>
<div class="line"><a class="code" href="apop_8h.html#ae32d4f48cff7788b617d40155703ae64">apop_name_add</a>(d-&gt;names, <span class="stringliteral">&quot;row 1&quot;</span>, <span class="charliteral">&#39;r&#39;</span>);</div>
<div class="line"><a class="code" href="apop_8h.html#ae32d4f48cff7788b617d40155703ae64">apop_name_add</a>(d-&gt;names, <span class="stringliteral">&quot;row 2&quot;</span>, <span class="charliteral">&#39;r&#39;</span>);</div>
<div class="line"><a class="code" href="apop_8h.html#ae32d4f48cff7788b617d40155703ae64">apop_name_add</a>(d-&gt;names, <span class="stringliteral">&quot;numeric column 0&quot;</span>, <span class="charliteral">&#39;c&#39;</span>);</div>
<div class="line"><a class="code" href="apop_8h.html#ae32d4f48cff7788b617d40155703ae64">apop_name_add</a>(d-&gt;names, <span class="stringliteral">&quot;text column 0&quot;</span>, <span class="charliteral">&#39;t&#39;</span>);</div>
<div class="line"><a class="code" href="apop_8h.html#ae32d4f48cff7788b617d40155703ae64">apop_name_add</a>(d-&gt;names, <span class="stringliteral">&quot;The name of the data set.&quot;</span>, <span class="charliteral">&#39;h&#39;</span>);</div>
<div class="line"></div>
<div class="line"><span class="comment">//or append several names at once</span></div>
<div class="line"><a class="code" href="apop_8h.html#aeca6159c82c21da0db82ae9ed0582dc6">apop_data_add_names</a>(d, <span class="charliteral">&#39;c&#39;</span>, <span class="stringliteral">&quot;numeric column 1&quot;</span>, <span class="stringliteral">&quot;numeric column 2&quot;</span>, <span class="stringliteral">&quot;numeric column 3&quot;</span>);</div>
<div class="line"></div>
<div class="line"><span class="comment">//point to element i from:</span></div>
<div class="line"></div>
<div class="line"><span class="keywordtype">char</span> *rowname_i = d-&gt;names-&gt;row[i];</div>
<div class="line"><span class="keywordtype">char</span> *colname_i = d-&gt;names-&gt;col[i];</div>
<div class="line"><span class="keywordtype">char</span> *textname_i = d-&gt;names-&gt;text[i];</div>
<div class="line"></div>
<div class="line"><span class="comment">//The vector also has a name:</span></div>
<div class="line"><span class="keywordtype">char</span> *vname = d-&gt;names-&gt;vector;</div>
</div><!-- fragment --><ul>
<li><a class="el" href="apop_8h.html#ae32d4f48cff7788b617d40155703ae64">apop_name_add()</a> : add one name </li>
<li><a class="el" href="apop_8h.html#aeca6159c82c21da0db82ae9ed0582dc6">apop_data_add_names()</a> : add a sequence of names at once </li>
<li><a class="el" href="apop_8h.html#ab4fd369dba75ec4782400cf7378e56ec">apop_name_stack()</a> : copy the contents of one name list to another </li>
<li><a class="el" href="apop_8h.html#adb1398925a93b7c2c502af8fb2abc49e">apop_name_find()</a> : find the row/col number for a given name. </li>
<li><a class="el" href="apop_8h.html#af7220e4b2fe43c56593af0ceb2c7001a">apop_name_print()</a> : print the <a class="el" href="structapop__name.html">apop_name</a> struct, for diagnostic purposes.</li>
</ul>
<p></div>
<p><h2><a class="anchor" name="textsec"><div class="trigger" onClick="showBranch('textsecd');swapFolder('textsecf')"><img src="right.png" border="0" id="textsecf" alt="pip"> Text data</div></a></h2><div class="branch" id="textsecd">
<p>The <a class="el" href="gentle.html#apop_data">apop_data</a> set includes a grid of strings, <code>text</code>, for holding text data.</p>
<p>Text should be encoded in UTF-8. US ASCII is a subset of UTF-8, so that's OK too.</p>
<p>There are a few simple forms for handling the <code>text</code> element of an <code><a class="el" href="structapop__data.html">apop_data</a></code> set, which handle the tedium of memory-handling for you.</p>
<ul>
<li>Use <a class="el" href="apop__data_8c.html#a9fba0f07c262a433133f6f3362617da3">apop_text_alloc</a> to allocate the block of text. It is actually a realloc function, which you can use to resize an existing block without leaks. </li>
<li>Use <a class="el" href="apop__data_8c.html#ac9c6f645acce82319fb85eb9d75112ba">apop_text_add</a> to add text elements. It replaces any existing text in the given slot without memory leaks. </li>
<li>The number of rows of text data in <code>tdata</code> is <code>tdata-&gt;textsize[0]</code>; the number of columns is <code>tdata-&gt;textsize[1]</code>. </li>
<li>Refer to individual elements using the usual 2-D array notation, <code>tdata-&gt;text[row][col]</code>. </li>
<li><code>x[0]</code> can always be written as <code>*x</code>, which may save some typing. The number of rows is <code>*tdata-&gt;textsize</code>. If you have a single column of text data (i.e., all data is in column zero), then item <code>i</code> is <code>*tdata-&gt;text[i]</code>. If you know you have exactly one cell of text, then its value is <code>**tdata-&gt;text</code>. </li>
<li>After <a class="el" href="apop__data_8c.html#a9fba0f07c262a433133f6f3362617da3">apop_text_alloc</a>, all elements are the empty string <code>""</code>, which you can check via <code>if (!strlen(dataset-&gt;text[i][j])) printf("&lt;blank&gt;")</code> or <code>if (!*dataset-&gt;text[i][j]) printf("&lt;blank&gt;")</code>. For the sake of efficiency when dealing with large, sparse data sets, all blank cells point to <em>the same</em> static empty string, meaning that freeing cells must be done with care. Your best bet is to rely on <a class="el" href="apop__data_8c.html#ac9c6f645acce82319fb85eb9d75112ba">apop_text_add</a>, <a class="el" href="apop__data_8c.html#a9fba0f07c262a433133f6f3362617da3">apop_text_alloc</a>, and <a class="el" href="apop__data_8c.html#a65386d7b37c2eaf37d1eb4e3ae7a7554">apop_text_free</a> to do the memory management for you.</li>
</ul>
<p>Here is a sample program that uses these forms, plus a few text-handling functions.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="apop_8h.html">apop.h</a>&gt;</span></div>
<div class="line"></div>
<div class="line"><span class="keywordtype">int</span> main(){</div>
<div class="line">    <a class="code" href="group__queries.html#ga236452b3305ea7b791c781d91a819e3b">apop_query</a>(<span class="stringliteral">&quot;create table data (name, city, state);&quot;</span></div>
<div class="line">            <span class="stringliteral">&quot;insert into data values (&#39;Mike Mills&#39;, &#39;Rockville&#39;, &#39;MD&#39;);&quot;</span></div>
<div class="line">            <span class="stringliteral">&quot;insert into data values (&#39;Bill Berry&#39;, &#39;Athens&#39;, &#39;GA&#39;);&quot;</span></div>
<div class="line">            <span class="stringliteral">&quot;insert into data values (&#39;Michael Stipe&#39;, &#39;Decatur&#39;, &#39;GA&#39;);&quot;</span>);</div>
<div class="line">    <a class="code" href="structapop__data.html">apop_data</a> *tdata = <a class="code" href="group__queries.html#ga2fbd580cd71f481ae04ee03b5cf64ff5">apop_query_to_text</a>(<span class="stringliteral">&quot;select name, city, state from data&quot;</span>);</div>
<div class="line">    printf(<span class="stringliteral">&quot;Customer #1: %s\n\n&quot;</span>, *tdata-&gt;text[0]);</div>
<div class="line"></div>
<div class="line">    printf(<span class="stringliteral">&quot;The data, via apop_data_print:\n&quot;</span>);</div>
<div class="line">    <a class="code" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print</a>(tdata);</div>
<div class="line"></div>
<div class="line">    <span class="comment">//the text alloc can be used as a text realloc:</span></div>
<div class="line">    <a class="code" href="apop_8h.html#a9fba0f07c262a433133f6f3362617da3">apop_text_alloc</a>(tdata, 1+tdata-&gt;textsize[0], tdata-&gt;textsize[1]);</div>
<div class="line">    <a class="code" href="apop_8h.html#ac9c6f645acce82319fb85eb9d75112ba">apop_text_add</a>(tdata, *tdata-&gt;textsize-1, 0, <span class="stringliteral">&quot;Peter Buck&quot;</span>);</div>
<div class="line">    <a class="code" href="apop_8h.html#ac9c6f645acce82319fb85eb9d75112ba">apop_text_add</a>(tdata, *tdata-&gt;textsize-1, 1, <span class="stringliteral">&quot;Berkeley&quot;</span>);</div>
<div class="line">    <a class="code" href="apop_8h.html#ac9c6f645acce82319fb85eb9d75112ba">apop_text_add</a>(tdata, *tdata-&gt;textsize-1, 2, <span class="stringliteral">&quot;CA&quot;</span>);</div>
<div class="line"></div>
<div class="line">    printf(<span class="stringliteral">&quot;\n\nAugmented data, printed via for loop:\n&quot;</span>);</div>
<div class="line">    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt; tdata-&gt;textsize[0]; i++){</div>
<div class="line">        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j=0; j&lt; tdata-&gt;textsize[1]; j++)</div>
<div class="line">            printf(<span class="stringliteral">&quot;%s\t&quot;</span>, tdata-&gt;text[i][j]);</div>
<div class="line">        printf(<span class="stringliteral">&quot;\n&quot;</span>);</div>
<div class="line">    }</div>
<div class="line"></div>
<div class="line">    <a class="code" href="structapop__data.html">apop_data</a> *states = <a class="code" href="apop_8h.html#a77c937ea4764d3da2ed9968273fa4544">apop_text_unique_elements</a>(tdata, 2);</div>
<div class="line">    <span class="keywordtype">char</span> *states_as_list = <a class="code" href="apop_8h.html#a6270ad55c566140e0d7962cb4d5daacc">apop_text_paste</a>(states, .between=<span class="stringliteral">&quot;, &quot;</span>);</div>
<div class="line">    printf(<span class="stringliteral">&quot;\n States covered: %s\n&quot;</span>, states_as_list);</div>
<div class="line">}</div>
</div><!-- fragment --><ul>
<li><a class="el" href="apop_8h.html#a82281b8673cda3645054d7e061bb1167">apop_data_transpose()</a> : also transposes the text data. Say that you use <code>dataset = apop_query_to_text("select onecolumn from data");</code> then you have a sequence of strings, <code>d-&gt;text[0][0], d-&gt;text[1][0], </code>.... After <code><a class="el" href="structapop__data.html">apop_data</a> dt = apop_data_transpose(dataset)</code>, you will have a single list of strings, <code>dt-&gt;text[0]</code>, which is often useful as input to list-of-strings handling functions.</li>
</ul>
<ul>
<li><a class="el" href="group__queries.html#ga2fbd580cd71f481ae04ee03b5cf64ff5">apop_query_to_text()</a> </li>
<li><a class="el" href="apop_8h.html#a9fba0f07c262a433133f6f3362617da3">apop_text_alloc()</a> : allocate or resize the text part of an <a class="el" href="gentle.html#apop_data">apop_data</a> set. </li>
<li><a class="el" href="apop_8h.html#ac9c6f645acce82319fb85eb9d75112ba">apop_text_add()</a>: replace a single cell of the text grid with new text. </li>
<li><a class="el" href="apop_8h.html#a6270ad55c566140e0d7962cb4d5daacc">apop_text_paste()</a> : convert a table of little strings into one long string. </li>
<li><a class="el" href="apop_8h.html#a77c937ea4764d3da2ed9968273fa4544">apop_text_unique_elements()</a> : get a sorted list of unique elements for one column of text. </li>
<li><a class="el" href="apop_8h.html#a65386d7b37c2eaf37d1eb4e3ae7a7554">apop_text_free()</a> : you may never need this, because <a class="el" href="apop_8h.html#a459acfde11f2c39f5c32cff377f85b9e">apop_data_free</a> calls it. </li>
<li><a class="el" href="apop_8h.html#aa9ec5e7ed223c09bec044ced54f24fad">apop_regex()</a> : friendlier front-end for POSIX-standard regular expression searching and pulling matches into a <a class="el" href="gentle.html#apop_data">apop_data</a> set. </li>
<li><a class="el" href="apop_8h.html#a6270ad55c566140e0d7962cb4d5daacc">apop_text_paste()</a>: produce a single string from a grid of text</li>
</ul>
<p></div>
<p><h2><a class="anchor" name="fact"><div class="trigger" onClick="showBranch('factd');swapFolder('factf')"><img src="right.png" border="0" id="factf" alt="pip"> Generating factors</div></a></h2><div class="branch" id="factd">
<p><em>Factor</em> is jargon for a numbered category. Number-crunching programs prefer integers over text, so we need a function to produce a one-to-one mapping from text categories into numeric factors.</p>
<p>A <em>dummy</em> is a variable that is either one or zero, depending on membership in a given group. Some methods (typically when the variable is an input or independent variable in a regression) prefer dummies; some methods (typically for outcome or dependent variables) prefer factors. The functions that generate factors and dummies will add an informational page to your <a class="el" href="gentle.html#apop_data">apop_data</a> set with a name like <code>&lt;categories for your_column&gt;</code> listing the conversion from the artificial numeric factor to the original data. Use <a class="el" href="apop__regression_8c.html#adec9170cf1382f6cd549a5819472eb51">apop_data_get_factor_names</a> to get a pointer to that page.</p>
<p>You can use the factor table to translate from numeric categories back to text (though you probably have the original text column in your data anyway).</p>
<p>This is especially useful if you want text in two data sets to have the same categories. Generate factors in the first set, then copy the factor list to the second, then run <a class="el" href="apop__regression_8c.html#a27405bd5fdf48d93c8918ca4c07ee260">apop_data_to_factors</a> on the second.</p>
<div class="fragment"><div class="line"><a class="code" href="apop_8h.html#a27405bd5fdf48d93c8918ca4c07ee260">apop_data_to_factors</a>(d1);</div>
<div class="line">d2-&gt;more = <a class="code" href="apop_8h.html#add67e6838ce2521a1d756894dc9b5ccf">apop_data_copy</a>(<a class="code" href="apop_8h.html#adec9170cf1382f6cd549a5819472eb51">apop_data_get_factor_names</a>(d1));</div>
<div class="line"><a class="code" href="apop_8h.html#a27405bd5fdf48d93c8918ca4c07ee260">apop_data_to_factors</a>(d2);</div>
</div><!-- fragment --><ul>
<li><a class="el" href="apop_8h.html#ae0c28224e10f8aacdaa2d130f96f6ba2">apop_data_to_dummies()</a> </li>
<li><a class="el" href="apop_8h.html#a27405bd5fdf48d93c8918ca4c07ee260">apop_data_to_factors()</a> </li>
<li><a class="el" href="apop_8h.html#adec9170cf1382f6cd549a5819472eb51">apop_data_get_factor_names()</a> </li>
<li><a class="el" href="apop_8h.html#a77c937ea4764d3da2ed9968273fa4544">apop_text_unique_elements()</a> </li>
<li><a class="el" href="apop_8h.html#a6d68bacc086420a382366f9118cf5bc7">apop_vector_unique_elements()</a></li>
</ul>
<p></div>
<p></div>
<p><h2><a class="anchor" name="dbs"><div class="trigger" onClick="showBranch('dbsd');swapFolder('dbsf')"><img src="right.png" border="0" id="dbsf" alt="pip"> Databases</div></a></h2><div class="branch" id="dbsd">
<p>These are convenience functions to handle interaction with SQLite or mySQL/mariaDB. They open one and only one database, and handle most of the interaction therewith for you.</p>
<p>You will probably first use <a class="el" href="group__conversions.html#gabd00e16791a970631c3ed4ac3487da46">apop_text_to_db</a> to pull data into the database, then <a class="el" href="group__queries.html#ga236452b3305ea7b791c781d91a819e3b">apop_query</a> to clean the data in the database, and finally <a class="el" href="group__queries.html#gade862993166c6482f53562fb77721db9">apop_query_to_data</a> to pull some subset of the data out for analysis.</p>
<p>See the <a class="el" href="db_moments.html">Database moments (plus pow()!)</a> page for not-SQL-standard math functions that you can use when sending queries from Apophenia, such as <code>pow</code>, <code>stddev</code>, or <code>sqrt</code>.</p>
<ul>
<li><a class="el" href="group__conversions.html#gabd00e16791a970631c3ed4ac3487da46">apop_text_to_db</a> : Read a text file on disk into the database. Most data analysis projects start with a call to this. </li>
<li><a class="el" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print</a> : If you include the argument <code>.output_type='d'</code>, this prints your <a class="el" href="gentle.html#apop_data">apop_data</a> set to the database. </li>
<li><a class="el" href="group__queries.html#ga236452b3305ea7b791c781d91a819e3b">apop_query</a> : Manipulate the database, return nothing (e.g., insert rows or create table). </li>
<li><a class="el" href="apop__db_8c.html#a4d81aff912df2982697038c51795e358">apop_db_open</a> : Optional, for when you want to use a database on disk. </li>
<li><a class="el" href="apop__db_8c.html#acf205f62e9432fe293b05c077d1b61d4">apop_db_close</a> : If you used <a class="el" href="apop__db_8c.html#a4d81aff912df2982697038c51795e358">apop_db_open</a>, you will need to use this too. </li>
<li><a class="el" href="apop__db_8c.html#a01d501ca84ade92bc72fd4f67a617f64">apop_table_exists</a> : Check to make sure you aren't reinventing or destroying data. Also, a clean way to drop a table.</li>
</ul>
<ul>
<li>Apophenia reserves the right to insert temp tables into the opened database. They will all have names beginning with "apop_", so the reader is advised to not use tables with such names, and is free to ignore or delete any such tables that turn up.</li>
</ul>
<ul>
<li>If you need to deal with two databases, use SQL's <a href="https://sqlite.org/lang_attach.html"><code>attach database</code></a>. By default with SQLite, Apophenia opens an in-memory database handle. It is a sensible workflow to use the faster in-memory database as the primary, and then attach an on-disk database to read in data and write final output tables.</li>
</ul>
<p><b>Extracting data from the database</b></p>
<ul>
<li><a class="el" href="apop_8h.html#aebe17fe6f2b0c49c0c90824b7a68ff90">apop_db_to_crosstab()</a>: take three columns in the database (row, column, value) and produce a table of values. </li>
<li><a class="el" href="group__queries.html#gade862993166c6482f53562fb77721db9">apop_query_to_data()</a> </li>
<li><a class="el" href="group__queries.html#ga1ad75f9dfba696a3741e28896ed5b75d">apop_query_to_float()</a> </li>
<li><a class="el" href="group__queries.html#ga1f06db34e05b5afb63112f83ad2317b8">apop_query_to_mixed_data()</a> </li>
<li><a class="el" href="group__queries.html#ga2fbd580cd71f481ae04ee03b5cf64ff5">apop_query_to_text()</a> </li>
<li><a class="el" href="group__queries.html#ga830703db838855b8abd636069c7145de">apop_query_to_vector()</a></li>
</ul>
<p><b>Writing data to the database</b></p>
<p>See the print functions below. E.g.</p>
<div class="fragment"><div class="line"><a class="code" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print</a>(yourdata, .output_type=<span class="charliteral">&#39;d&#39;</span>, .output_name=<span class="stringliteral">&quot;dbtab&quot;</span>);</div>
</div><!-- fragment --><p><h2><a class="anchor" name="cmdline"><div class="trigger" onClick="showBranch('cmdlined');swapFolder('cmdlinef')"><img src="right.png" border="0" id="cmdlinef" alt="pip"> Command-line utilities</div></a></h2><div class="branch" id="cmdlined">
<p>A few functions have proven to be useful enough to be worth breaking out into their own programs, for use in scripts or other data analysis from the command line:</p>
<ul>
<li>The <code>apop_text_to_db</code> command line utility is a wrapper for the <a class="el" href="group__conversions.html#gabd00e16791a970631c3ed4ac3487da46">apop_text_to_db</a> command. </li>
<li>The <code>apop_db_to_crosstab</code> function is a wrapper for the <a class="el" href="apop__conversions_8c.html#aebe17fe6f2b0c49c0c90824b7a68ff90">apop_db_to_crosstab</a> function.</li>
</ul>
<p></div>
<p></div>
<p><h2><a class="anchor" name="Modesec"><div class="trigger" onClick="showBranch('Modesecd');swapFolder('Modesecf')"><img src="right.png" border="0" id="Modesecf" alt="pip"> Models</div></a></h2><div class="branch" id="Modesecd">
<p>This segment discusses the use of existing <a class="el" href="structapop__model.html">apop_model</a> objects. If you need to write a new model, see <a class="el" href="modeldetails.html">Writing new models</a>.</p>
<p><h2><a class="anchor" name="introtomodels"><div class="trigger" onClick="showBranch('introtomodelsd');swapFolder('introtomodelsf')"><img src="right.png" border="0" id="introtomodelsf" alt="pip"> Introduction</div></a></h2><div class="branch" id="introtomodelsd">
<p>Begin with the most common use: the <code>estimate</code> function will estimate the parameters of your model. Just prep the data, select a model, and produce an estimate:</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *data = <a class="code" href="group__queries.html#gade862993166c6482f53562fb77721db9">apop_query_to_data</a>(<span class="stringliteral">&quot;select outcome, in1, in2, in3 from dataset&quot;</span>);</div>
<div class="line"><a class="code" href="structapop__model.html">apop_model</a> *the_estimate = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(data, <a class="code" href="group__models.html#ga319b8e3d5025a6fc94829a978273eb8b">apop_probit</a>);</div>
<div class="line"><a class="code" href="group__output.html#gab7bf9adff9b97c60c5839ae72ba4fb7f">apop_model_print</a>(the_estimate, NULL);</div>
</div><!-- fragment --><p>Along the way to estimating the parameters, most models also find covariance estimates for the parameters, calculate statistics like log likelihood, and so on, which the final print statement will show.</p>
<p>The <code>apop_probit</code> model that ships with Apophenia is unparameterized: <code>apop_probit.parameters==NULL</code>. The output from the estimation, <code>the_estimate</code>, has the same form as <code>apop_probit</code>, but <code>the_estimate-&gt;parameters</code> has a meaningful value.</p>
<p><h2><a class="anchor" name="covandstuff"><div class="trigger" onClick="showBranch('covandstuffd');swapFolder('covandstufff')"><img src="right.png" border="0" id="covandstufff" alt="pip"> More estimation output</div></a></h2><div class="branch" id="covandstuffd">
<p>A call to <a class="el" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a> produces more than just the estimated parameters. Most will produce any of a covariance matrix, some hypothesis tests, a list of expected values, log likelihood, AIC, AIC_c, BIC, et cetera.</p>
<p>First, note that if you don't want all that, adding to your model an <a class="el" href="structapop__parts__wanted__settings.html">apop_parts_wanted_settings</a> group with its default values (see below on settings groups) signals to the model that you want only the parameters and to not waste CPU time on covariances, expected values, et cetera. See the <a class="el" href="structapop__parts__wanted__settings.html">apop_parts_wanted_settings</a> documentation for examples and further refinements.</p>
<ul>
<li>The actual parameter estimates are in an <a class="el" href="gentle.html#apop_data">apop_data</a> set at <code>your_model-&gt;parameters</code>.</li>
</ul>
<ul>
<li>A pointer to the <a class="el" href="gentle.html#apop_data">apop_data</a> set used for estimation, named <code>data</code>.</li>
</ul>
<ul>
<li>Scalar statistics of the model are listed in the output model's <code>info</code> group, and can be retrieved via a form like</li>
</ul>
<div class="fragment"><div class="line"><a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(your_model-&gt;info, .rowname=<span class="stringliteral">&quot;log likelihood&quot;</span>);</div>
<div class="line"><span class="comment">//or</span></div>
<div class="line"><a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(your_model-&gt;info, .rowname=<span class="stringliteral">&quot;AIC&quot;</span>);</div>
</div><!-- fragment --><ul>
<li>Covariances of the parameters are a page appended to the parameters; retrieve via</li>
</ul>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *cov = <a class="code" href="apop_8h.html#a044771db76f0afb410a8d4f05ac51ca3">apop_data_get_page</a>(your_model-&gt;parameters, <span class="stringliteral">&quot;&lt;Covariance&gt;&quot;</span>);</div>
</div><!-- fragment --><ul>
<li>If the model calculates it, the table of expected values (typically including expected value, actual value, and residual) is a page stapled to the main info page. This is mostly for regression-type models. Retrieve via:</li>
</ul>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *predict = <a class="code" href="apop_8h.html#a044771db76f0afb410a8d4f05ac51ca3">apop_data_get_page</a>(your_model-&gt;info, <span class="stringliteral">&quot;&lt;Predicted&gt;&quot;</span>);</div>
</div><!-- fragment --><p></div>
<p>But we expect much more from a model than just estimating parameters from data.</p>
<p>Continuing the above example where we got an estimated Probit model named <code>the_estimate</code>, we can interrogate the estimate in various familiar ways. In each of the following examples, the model object holds enough information that the generic function being called can do its work:</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> *expected_value = <a class="code" href="group__models.html#ga867f0b8df261866bb18e9040871c51f0">apop_predict</a>(NULL, the_estimate);</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> density_under =  <a class="code" href="group__models.html#gaee16258d3e52671e254f3ce03bca7754">apop_cdf</a>(expected_value, the_estimate);</div>
<div class="line"></div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> *draws = <a class="code" href="apop_8h.html#a4e204cdc7429bef3ca2efb83453fe77a">apop_model_draws</a>(the_estimate, .count=1000);</div>
</div><!-- fragment --><p>Apophenia ships with many well-known models for your immediate use, including probability distributions, such as the <a class="el" href="group__models.html#ga297c2dc8dbdf4ae93fc7e10860400466">apop_normal</a>, <a class="el" href="group__models.html#ga9c67838a4761452cc2c9113557f2a070">apop_poisson</a>, or <a class="el" href="group__models.html#gac11decb2971532ada7d2cd4982848b90">apop_beta</a> models. The data is assumed to have been drawn from a given distribution and the question is only what distributional parameters best fit; e.g., assume the data is Normally distributed and find the mean and variance: <a class="el" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(<code>your_data</code>, <a class="el" href="group__models.html#ga297c2dc8dbdf4ae93fc7e10860400466">apop_normal</a>).</p>
<p>There are also linear models like <a class="el" href="group__models.html#ga85524b1deabd5829227888fd91740cc3">apop_ols</a>, <a class="el" href="group__models.html#ga319b8e3d5025a6fc94829a978273eb8b">apop_probit</a>, and <a class="el" href="group__models.html#gaa88a402cf7d02c25c4613a74ace1b0c1">apop_logit</a>. As in the example, they are on equal footing with the distributions, so nothing keeps you from making random draws from an estimated linear model.</p>
<ul>
<li>If you send a data set with the <code>weights</code> vector filled, <a class="el" href="group__models.html#ga85524b1deabd5829227888fd91740cc3">apop_ols</a> estimates Weighted OLS.</li>
</ul>
<ul>
<li>If the dependent variable has more than categories, The <a class="el" href="group__models.html#ga319b8e3d5025a6fc94829a978273eb8b">apop_probit</a> and <a class="el" href="group__models.html#gaa88a402cf7d02c25c4613a74ace1b0c1">apop_logit</a> models estimate a multinomial logit or probit.</li>
</ul>
<ul>
<li>There are separate <a class="el" href="group__models.html#ga297c2dc8dbdf4ae93fc7e10860400466">apop_normal</a> and <a class="el" href="group__models.html#ga3a4389e6e2a1a2ea606f89d55ec558d3">apop_multivariate_normal</a> functions because the parameter formats are slightly different: the univariate Normal keeps both μ and σ in the vector element of the parameters; the multivariate version uses the vector for the vector of means and the matrix for the Σ matrix. The univariate is so heavily used that it merits a special-case model.</li>
</ul>
<p>Simulation models seem to not fit this form, but you will see below that if you can write an objective function for the <code>p</code> method of the model, you can use the above tools. Notably, you can estimate parameters via maximum likelihood and then give confidence intervals around those parameters.</p>
<p>But some models have to break uniformity, like how a histogram has a list of bins that makes no sense for a Normal distribution. These are held in <em>settings groups</em>, which you will occasionally need to tweak to modify how a model is handled or estimated. The most common example would be for maximum likelihood, eg.</p>
<div class="fragment"><div class="line"><span class="comment">//Probit uses MLE. Redo the estimation using Newton&#39;s Method</span></div>
<div class="line"><a class="code" href="group__settings.html#ga8fb1877a3cc29edd685a68dd6b4a35dc">Apop_settings_add_group</a>(the_estimate, apop_mle, .verbose=<span class="charliteral">&#39;y&#39;</span>, </div>
<div class="line">                        .tolerance=1e-4, .method=APOP_RF_NEWTON);</div>
<div class="line"><a class="code" href="structapop__model.html">apop_model</a> *re_est = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(data, the_estimate);</div>
</div><!-- fragment --><p>See below for more details on using settings groups.</p>
<p><h2><a class="anchor" name="modelparameterization"><div class="trigger" onClick="showBranch('modelparameterizationd');swapFolder('modelparameterizationf')"><img src="right.png" border="0" id="modelparameterizationf" alt="pip"> Parameterizing or initializing a model</div></a></h2><div class="branch" id="modelparameterizationd">
<p>The models that ship with Apophenia have the requisite procedures for estimation, making draws, and so on, but have <code>parameters==NULL</code> and <code>settings==NULL</code>. The model is thus, for many purposes, incomplete, and you will need to take some action to complete the model. As per the examples to follow, there are several possibilities:</p>
<ul>
<li>Estimate it! Almost all models can be sent with a data set as an argument to the <code>apop_estimate</code> function. The input model is unchanged, but the output model has parameters and settings in place.</li>
</ul>
<ul>
<li>If your model has a fixed number of numeric parameters, then you can set them with <a class="el" href="group__models.html#ga36cae46a2950eb6cfb65794cdd5dac7e">apop_model_set_parameters</a>.</li>
</ul>
<ul>
<li>If your model has a variable number of parameters, you can directly set the <code>parameters</code> element via <code>apop_data_falloc</code>. For most purposes, you will also need to set the <code>msize1</code>, <code>msize2</code>, <code>vsize</code>, and <code>dsize</code> elements to the size you want. See the example below.</li>
</ul>
<ul>
<li>Some models have disparate, non-numeric settings rather than a simple matrix of parameters. For example, an kernel density estimate needs a model as a kernel and a base data set, which can be set via <a class="el" href="group__settings.html#ga8e5f7950ae47b299e20f2c447199cf6f">apop_model_copy_set</a>.</li>
</ul>
<p>Here is an example that shows the options for parameterizing a model. After each parameterization, 20 draws are made and written to a file named draws-[modelname].</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="apop_8h.html">apop.h</a>&gt;</span></div>
<div class="line"></div>
<div class="line"><span class="preprocessor">#define print_draws(mm) apop_data_print(apop_model_draws(mm, 20),\</span></div>
<div class="line"><span class="preprocessor">                                        .output_name= &quot;draws-&quot; #mm);</span></div>
<div class="line"></div>
<div class="line"><span class="keywordtype">int</span> main(){</div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *uniform_20 = <a class="code" href="group__models.html#ga36cae46a2950eb6cfb65794cdd5dac7e">apop_model_set_parameters</a>(<a class="code" href="group__models.html#ga95be4cc3931072118a7c8ffece73fbe1">apop_uniform</a>, 0, 20);</div>
<div class="line">    <a class="code" href="structapop__data.html">apop_data</a> *d = <a class="code" href="apop_8h.html#a4e204cdc7429bef3ca2efb83453fe77a">apop_model_draws</a>(uniform_20, 10);</div>
<div class="line"></div>
<div class="line">    <span class="comment">//Estimate a Normal distribution from the data:</span></div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *N = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(d, <a class="code" href="group__models.html#ga297c2dc8dbdf4ae93fc7e10860400466">apop_normal</a>);</div>
<div class="line">    print_draws(N);</div>
<div class="line"></div>
<div class="line">    <span class="comment">//estimate a one-dimensional multivariate Normal from the data:</span></div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *mvN = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(d, <a class="code" href="group__models.html#ga3a4389e6e2a1a2ea606f89d55ec558d3">apop_multivariate_normal</a>);</div>
<div class="line">    print_draws(mvN);</div>
<div class="line"></div>
<div class="line"></div>
<div class="line">    <span class="comment">//fixed parameter list:</span></div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *std_normal = <a class="code" href="group__models.html#ga36cae46a2950eb6cfb65794cdd5dac7e">apop_model_set_parameters</a>(<a class="code" href="group__models.html#ga297c2dc8dbdf4ae93fc7e10860400466">apop_normal</a>, 0, 1);</div>
<div class="line">    print_draws(std_normal);</div>
<div class="line"></div>
<div class="line">    <span class="comment">//variable-size parameter list:</span></div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *std_multinormal = <a class="code" href="group__models.html#ga87d3a65f43071b9d61d9edd59249c629">apop_model_copy</a>(<a class="code" href="group__models.html#ga3a4389e6e2a1a2ea606f89d55ec558d3">apop_multivariate_normal</a>);</div>
<div class="line">    std_multinormal-&gt;msize1 =</div>
<div class="line">    std_multinormal-&gt;msize2 =</div>
<div class="line">    std_multinormal-&gt;vsize =</div>
<div class="line">    std_multinormal-&gt;<a class="code" href="structapop__model.html#a5e31343782fe36a70e60e6d0808c64b5">dsize</a> = 3;</div>
<div class="line">    std_multinormal-&gt;<a class="code" href="structapop__model.html#ae016a9e13725e84a1188bc81c8f09e45">parameters</a> = <a class="code" href="apop_8h.html#acb6b2c53478f9db902ba0340ba499819">apop_data_falloc</a>((3, 3, 3),</div>
<div class="line">                                1,  1, 0, 0, </div>
<div class="line">                                1,  0, 1, 0,</div>
<div class="line">                                1,  0, 0, 1);</div>
<div class="line">    print_draws(std_multinormal);</div>
<div class="line"></div>
<div class="line"></div>
<div class="line">    <span class="comment">//estimate a KDE using the defaults:</span></div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *k = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(d, <a class="code" href="group__models.html#gaded18b6bf8bd63a53f7b72ed5f27fdb5">apop_kernel_density</a>);</div>
<div class="line">    print_draws(k);</div>
<div class="line"></div>
<div class="line">    <span class="comment">/*the documentation tells us that a KDE estimation consists of filling </span></div>
<div class="line"><span class="comment">      an apop_kernel_density_settings group, so we can set it to use a </span></div>
<div class="line"><span class="comment">      Normal(μ, 2) kernel via: */</span></div>
<div class="line"></div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *k2 = <a class="code" href="group__settings.html#ga8e5f7950ae47b299e20f2c447199cf6f">apop_model_copy_set</a>(<a class="code" href="group__models.html#gaded18b6bf8bd63a53f7b72ed5f27fdb5">apop_kernel_density</a>, <a class="code" href="group__models.html#gaded18b6bf8bd63a53f7b72ed5f27fdb5">apop_kernel_density</a>, </div>
<div class="line">                         .base_data=d,</div>
<div class="line">                         .kernel = <a class="code" href="group__models.html#ga36cae46a2950eb6cfb65794cdd5dac7e">apop_model_set_parameters</a>(<a class="code" href="group__models.html#ga297c2dc8dbdf4ae93fc7e10860400466">apop_normal</a>, 0, 2));</div>
<div class="line">    print_draws(k2);</div>
<div class="line">}</div>
</div><!-- fragment --><p></div>
<p>Where to from here? See the <a class="el" href="group__models.html">Models</a> page for a list of basic functions that make use of them, along with a list of the canned models, including popular favorites like <a class="el" href="group__models.html#gac11decb2971532ada7d2cd4982848b90">apop_beta</a>, <a class="el" href="group__models.html#ga20017e88fda3695f093279c8191f48e3">apop_binomial</a>, <a class="el" href="group__models.html#ga1efcbbe0410c89401354c762363a909f">apop_iv</a> (instrumental variables), <a class="el" href="group__models.html#gaded18b6bf8bd63a53f7b72ed5f27fdb5">apop_kernel_density</a>, <a class="el" href="group__models.html#gaf7d01cc4076c9319036a0cce882dac62">apop_loess</a>, <a class="el" href="group__models.html#ga4a917e02f41e9c6a0ad5e084e3ca42d2">apop_lognormal</a>, <a class="el" href="group__models.html#gaca8f3323c57e0223a9f3c0f991c9760e">apop_pmf</a> (see the histogram section below), and <a class="el" href="group__models.html#ga9c67838a4761452cc2c9113557f2a070">apop_poisson</a>.</p>
<p>If you need to write a new model, see <a class="el" href="modeldetails.html">Writing new models</a>.</p>
<p></div>
<p><h2><a class="anchor" name="mathmethods"><div class="trigger" onClick="showBranch('mathmethodsd');swapFolder('mathmethodsf')"><img src="right.png" border="0" id="mathmethodsf" alt="pip"> Model methods</div></a></h2><div class="branch" id="mathmethodsd">
<ul>
<li><a class="el" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate()</a> : estimate the parameters of the model with data. </li>
<li><a class="el" href="group__models.html#ga867f0b8df261866bb18e9040871c51f0">apop_predict()</a> : the expected value function. </li>
<li><a class="el" href="group__models.html#ga87f15cd1923ed7b5b46e52a77ca1dfef">apop_draw()</a> : random draws from an estimated model. </li>
<li><a class="el" href="group__models.html#ga2510c7df8ccd834732ee208005738b0f">apop_p()</a> : the probability of a given data set given the model. </li>
<li><a class="el" href="group__models.html#ga5e160c8d0a0ee660c89fb3f6de2af89c">apop_log_likelihood()</a> : the log of <a class="el" href="group__models.html#ga2510c7df8ccd834732ee208005738b0f">apop_p</a> </li>
<li><a class="el" href="group__models.html#ga228b013bc80f308900882312cab93eb3">apop_score()</a> : the derivative of <a class="el" href="group__models.html#ga5e160c8d0a0ee660c89fb3f6de2af89c">apop_log_likelihood</a> </li>
<li><a class="el" href="group__output.html#gab7bf9adff9b97c60c5839ae72ba4fb7f">apop_model_print()</a> : write to screen, file, or database</li>
</ul>
<ul>
<li><a class="el" href="group__models.html#ga87d3a65f43071b9d61d9edd59249c629">apop_model_copy()</a> : duplicate a model </li>
<li><a class="el" href="group__models.html#ga36cae46a2950eb6cfb65794cdd5dac7e">apop_model_set_parameters()</a> : Models ship with no parameters set. Use this to convert a Normal(μ, σ) with unknown μ and σ into a Normal(0, 1), for example. </li>
<li><a class="el" href="group__models.html#gab29755a86b82dfe31cdd01d55431cc09">apop_model_free()</a> </li>
<li><a class="el" href="group__models.html#gaa86381a72949fd3ffafefa87e3f51a4f">apop_model_clear()</a>, <a class="el" href="group__models.html#ga373ccc4cb87b162d62d07a2685ccd8ba">apop_prep()</a> : remove the parameters from a parameterized model. Used infrequently. </li>
<li><a class="el" href="apop_8h.html#a4e204cdc7429bef3ca2efb83453fe77a">apop_model_draws()</a> : many random draws from an estimated model.</li>
</ul>
<p></div>
<p><h2><a class="anchor" name="Update"><div class="trigger" onClick="showBranch('Updated');swapFolder('Updatef')"><img src="right.png" border="0" id="Updatef" alt="pip"> Filtering &amp; updating</div></a></h2><div class="branch" id="Updated">
<p>The model structure makes it easy to generate new models that are variants of prior models. Bayesian updating, for example, takes in one <a class="el" href="structapop__model.html">apop_model</a> that we call the prior, one <a class="el" href="structapop__model.html">apop_model</a> that we call a likelihood, and outputs an <a class="el" href="structapop__model.html">apop_model</a> that we call the posterior. One can produce complex models using simpler transformations as well. For example, to generate a one-parameter Normal(μ, 1) given the code for for a Normal(μ, σ):</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__model.html">apop_model</a> *N_sigma1 = <a class="code" href="apop_8h.html#ab4f6ca7e042ac7c09c3c01d29702a739">apop_model_fix_params</a>(<a class="code" href="group__models.html#ga36cae46a2950eb6cfb65794cdd5dac7e">apop_model_set_parameters</a>(<a class="code" href="group__models.html#ga297c2dc8dbdf4ae93fc7e10860400466">apop_normal</a>, NAN, 1));</div>
</div><!-- fragment --><p>This can be used anywhere the original Normal distribution can be. If we need to truncate the distribution in the data space:</p>
<div class="fragment"><div class="line"><span class="comment">//The constraint function.</span></div>
<div class="line"><span class="keywordtype">double</span> over_zero(<a class="code" href="structapop__data.html">apop_data</a> *in, <a class="code" href="structapop__model.html">apop_model</a> *m){</div>
<div class="line">    <span class="keywordflow">return</span> <a class="code" href="group__data__set__get.html#ga06107779e4a3eb229715ceca24352b89">apop_data_get</a>(in) &gt; 0;</div>
<div class="line">}</div>
<div class="line"></div>
<div class="line"><a class="code" href="structapop__model.html">apop_model</a> *trunc = <a class="code" href="apop_8h.html#a34644b9d55c7528b47f0b887c9241928">apop_model_dconstrain</a>(.base_model=N_sigma1,</div>
<div class="line">                                          .constraint=over_zero);</div>
</div><!-- fragment --><p>Chaining together simpler transformations is an easy method to produce models of arbitrary detail.</p>
<ul>
<li><a class="el" href="apop_8h.html#ac7c0ddb63673ddd40ba69f7dc8bff800">apop_update()</a> : Bayesian updating </li>
<li><a class="el" href="apop_8h.html#a15d9e73643bf66f7d69c7ebf959757aa">apop_model_coordinate_transform()</a> : apply an invertible transformation to the data space </li>
<li><a class="el" href="apop_8h.html#a34644b9d55c7528b47f0b887c9241928">apop_model_dconstrain()</a> : constrain the data space of a model to a subspace. E.g., truncate a Normal distribution so <img class="formulaInl" alt="$x>0$" src="form_103.png"/>. </li>
<li><a class="el" href="apop_8h.html#ab4f6ca7e042ac7c09c3c01d29702a739">apop_model_fix_params()</a> : hold some parameters constant </li>
<li><a class="el" href="apop_8h.html#a40bb98b02fde8f8e4ab59a4d127b692f">apop_model_mixture()</a> : a linear combination of models </li>
<li>apop_model_stack() : If <img class="formulaInl" alt="$(p_1, p_2)$" src="form_104.png"/> has a Normal distribution and <img class="formulaInl" alt="$p_3$" src="form_105.png"/> has an independent Poisson distribution, then <img class="formulaInl" alt="$(p_1, p_2, p_3)$" src="form_106.png"/> has an <code>apop_model_stack(apop_normal, apop_poisson)</code> distribution. </li>
<li><a class="el" href="apop_8h.html#ab664f68b5e481bec56c8ddc7f16928c0">apop_model_dcompose()</a> : use the output of one model as a data set for another</li>
</ul>
<p></div>
<p><h2><a class="anchor" name="modelsettings"><div class="trigger" onClick="showBranch('modelsettingsd');swapFolder('modelsettingsf')"><img src="right.png" border="0" id="modelsettingsf" alt="pip"> Settings groups</div></a></h2><div class="branch" id="modelsettingsd">
<p>[For info on specific settings groups and their contents and use, see the <a class="el" href="group__settings.html">Settings</a> page.]</p>
<p>Describing a statistical, agent-based, social, or physical model in a standardized form is difficult because every model has significantly different settings. E.g., an MLE requires a method of search (conjugate gradient, simplex, simulated annealing), and a histogram needs the number of bins to be filled with data.</p>
<p>So, the <a class="el" href="structapop__model.html">apop_model</a> includes a single list which can hold an arbitrary number of groups of settings, like the search specifications for finding the maximum likelihood, a histogram for making random draws, and options about the model type.</p>
<p>Settings groups are automatically initialized with default values when needed. If the defaults do no harm, then you don't need to think about these settings groups at all.</p>
<p>Here is an example where a settings group is worth tweaking: the <a class="el" href="structapop__parts__wanted__settings.html">apop_parts_wanted_settings</a> group indicates which parts of the auxiliary data you want.</p>
<div class="fragment"><div class="line">1 <a class="code" href="structapop__model.html">apop_model</a> *m = <a class="code" href="group__models.html#ga87d3a65f43071b9d61d9edd59249c629">apop_model_copy</a>(<a class="code" href="group__models.html#ga85524b1deabd5829227888fd91740cc3">apop_ols</a>);</div>
<div class="line">2 <a class="code" href="group__settings.html#ga8fb1877a3cc29edd685a68dd6b4a35dc">Apop_settings_add_group</a>(m, apop_parts_wanted, .covariance=<span class="charliteral">&#39;y&#39;</span>);</div>
<div class="line">3 <a class="code" href="structapop__model.html">apop_model</a> *est = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(data, m);</div>
</div><!-- fragment --><p>Line one establishes the baseline form of the model. Line two adds a settings group of type <a class="el" href="structapop__parts__wanted__settings.html">apop_parts_wanted_settings</a> to the model. By default other auxiliary items, like the expected values, are set to <code>'n'</code> when using this group, so this specifies that we want covariance and only covariance. Having stated our preferences, line three does the estimation we want.</p>
<p>Notice that the <code>_settings</code> ending to the settings group's name isn't written&mdash;macros make it happen. The remaining arguments to <code>Apop_settings_add_group</code> (if any) follow the <a class="el" href="designated.html">Designated initializers</a> syntax of the form <code>.setting=value</code>.</p>
<p>There is an <a class="el" href="group__settings.html#ga8e5f7950ae47b299e20f2c447199cf6f">apop_model_copy_set</a> macro that adds a settings group when it is first copied, joining up lines one and two above:</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__model.html">apop_model</a> *m = <a class="code" href="group__settings.html#ga8e5f7950ae47b299e20f2c447199cf6f">apop_model_copy_set</a>(<a class="code" href="group__models.html#ga85524b1deabd5829227888fd91740cc3">apop_ols</a>, apop_parts_wanted, .covariance=<span class="charliteral">&#39;y&#39;</span>);</div>
</div><!-- fragment --><p>Settings groups are copied with the model, which facilitates chaining estimations. Continuing the above example, you could re-estimate to get the predicted values and covariance via:</p>
<div class="fragment"><div class="line"><a class="code" href="group__settings.html#ga8659282b6003f7c2adac6f4baad5fae4">Apop_settings_set</a>(est, apop_parts_wanted, predicted, <span class="charliteral">&#39;y&#39;</span>);</div>
<div class="line"><a class="code" href="structapop__model.html">apop_model</a> *est2 = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(data, est);</div>
</div><!-- fragment --><ul>
<li><a class="el" href="group__settings.html#ga8659282b6003f7c2adac6f4baad5fae4">Apop_settings_set</a>, for modifying a single setting, doesn't use the designated initializers format.</li>
</ul>
<ul>
<li>The <a class="el" href="group__settings.html">Settings</a> page lists the settings structures included in Apophenia and their use.</li>
</ul>
<ul>
<li>Because the settings groups are buried within the model, debugging them can be a pain. Here is a documented macro for <code>gdb</code> that will help you pull a settings group out of a model for your inspection, to cut and paste into your <code></code>.gdbinit. It shouldn't be too difficult to modify this macro for other debuggers.</li>
</ul>
<div class="fragment"><div class="line">define get_group</div>
<div class="line">    set $group = ($arg1_settings *) apop_settings_get_grp( $arg0, <span class="stringliteral">&quot;$arg1&quot;</span>, 0 )</div>
<div class="line">    p *$group</div>
<div class="line">end</div>
<div class="line">document get_group </div>
<div class="line">Gets a settings group from a model.</div>
<div class="line">Give the model name and the name of the group, like</div>
<div class="line">get_group my_model apop_mle </div>
<div class="line">and I will set a gdb variable named $group that points to that model, </div>
<div class="line">which you can use like any other pointer. For example, print the contents with</div>
<div class="line">p *$group</div>
<div class="line">The contents of $group are printed to the screen as visible output to <span class="keyword">this</span> macro.</div>
<div class="line">end </div>
</div><!-- fragment --><p>For just using a model, that's all of what you need to know. For details on writing a new settings group, see <a class="el" href="modeldetails.html#settingswriting">Writing new settings groups</a> .</p>
<ul>
<li><a class="el" href="group__settings.html#ga8fb1877a3cc29edd685a68dd6b4a35dc">Apop_settings_add_group</a> </li>
<li><a class="el" href="group__settings.html#ga8659282b6003f7c2adac6f4baad5fae4">Apop_settings_set</a> </li>
<li><a class="el" href="group__settings.html#ga4ba53402807c9e6b82254fb552029501">Apop_settings_get</a> get a single element from a settings group. </li>
<li><a class="el" href="group__settings.html#gaafba071ab93465fed4cac1020082b4c8">Apop_settings_get_group</a> get the whole settings group.</li>
</ul>
<p></div>
<p></div>
<p></div>
<p><h2><a class="anchor" name="Test"><div class="trigger" onClick="showBranch('Testd');swapFolder('Testf')"><img src="right.png" border="0" id="Testf" alt="pip"> Tests &amp; diagnostics</div></a></h2><div class="branch" id="Testd">
<p>Just about any hypothesis test consists of a few common steps:</p>
<ul>
<li>specify a statistic </li>
<li>State the statistic's hypothesized distribution </li>
<li>Find the odds that the statistic would lie within some given range, like <em>greater than zero</em> or <em>near 1.1</em>.</li>
</ul>
<p>If the statistic is from a common form, like the parameters from an OLS regression, then the commonly-associated <img class="formulaInl" alt="$t$" src="form_44.png"/> test is probably included as part of the estimation output, typically as a row in the <code>info</code> element of the <a class="el" href="gentle.html#apop_data">apop_data</a> set.</p>
<p>Some tests, like ANOVA, produce a statistic using a specialized procedure, so Apophenia includes some functions, like <a class="el" href="group__asst__tests.html#ga12b92b7494f4e24d2f47244e3792d566">apop_test_anova_independence</a> and <a class="el" href="group__histograms.html#ga98ccaa9de7cb16e60a435c56eaf3400d">apop_test_kolmogorov</a>, to produce the statistic and look up its significance level.</p>
<p>If you are producing a statistic that you know has a common form, like a central limit theorem tells you that your statistic is Normally distributed, then the convenience function <a class="el" href="apop__tests_8c.html#a650d1ecf6f78c3bdb85c8fce71dfae05">apop_test</a> will do the final lookup step of checking where your statistic lies on your chosen distribution.</p>
<ul>
<li><a class="el" href="apop_8h.html#a650d1ecf6f78c3bdb85c8fce71dfae05">apop_test()</a> </li>
<li><a class="el" href="group__ttest.html#ga0ae576fb823e38f7b6f2ce1892e4b014">apop_paired_t_test()</a> </li>
<li><a class="el" href="apop_8h.html#a1902b3fd6a6682d1453a80b788e44ebb">apop_f_test()</a> </li>
<li><a class="el" href="group__ttest.html#ga65c94b79fe91308dc97829741acc9322">apop_t_test()</a> </li>
<li><a class="el" href="group__asst__tests.html#ga12b92b7494f4e24d2f47244e3792d566">apop_test_anova_independence()</a> </li>
<li><a class="el" href="apop_8h.html#a8627a1eebf904699e90e66adc8b4b021">apop_test_fisher_exact()</a> </li>
<li><a class="el" href="group__histograms.html#ga98ccaa9de7cb16e60a435c56eaf3400d">apop_test_kolmogorov()</a> </li>
<li><a class="el" href="group__regression.html#gaceead3e4aaba543b7fc1d194fcbc40f8">apop_estimate_coefficient_of_determination()</a> </li>
<li><a class="el" href="group__regression.html#ga33c353f768fb4b0e7a3f98ab7a546ba5">apop_estimate_r_squared()</a> </li>
<li><a class="el" href="apop_8h.html#ae165633525c5272ab83794b60076eddc">apop_estimate_parameter_tests()</a></li>
</ul>
<p>See also the example at the end of <a class="el" href="gentle.html">A quick overview</a>.</p>
<p><h2><a class="anchor" name="Mont"><div class="trigger" onClick="showBranch('Montd');swapFolder('Montf')"><img src="right.png" border="0" id="Montf" alt="pip"> Monte Carlo methods</div></a></h2><div class="branch" id="Montd">
<ul>
<li><a class="el" href="apop_8h.html#a4eb57b1f864c650ed6d7e029550e7bb8">apop_bootstrap_cov()</a> </li>
<li><a class="el" href="apop_8h.html#a8a85afbb18d256bcc8942d2eca7470ed">apop_jackknife_cov()</a></li>
</ul>
<p></div>
<p>To give another example of testing, here is a function that used to be a part of Apophenia, but seemed a bit out of place. Here it is as a sample:</p>
<div class="fragment"><div class="line"><span class="comment">// Input: any old vector. Output: 1 - the p-value for a chi-squared</span></div>
<div class="line"><span class="comment">// test to answer the question, &quot;with what confidence can I reject the</span></div>
<div class="line"><span class="comment">// hypothesis that the variance of my data is zero?&quot;</span></div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> apop_test_chi_squared_var_not_zero(<span class="keyword">const</span> gsl_vector *in){</div>
<div class="line">    <a class="code" href="apop_8h.html#abd17601e642688d033f62d346f3dace6">Apop_stopif</a>(!in, <span class="keywordflow">return</span> NAN, 0, <span class="stringliteral">&quot;input vector is NULL. Doing nothing.&quot;</span>);</div>
<div class="line">    <span class="keywordtype">double</span>          sum=0;</div>
<div class="line">    gsl_vector *normed = <a class="code" href="apop_8h.html#a194ad92b1bc632ba1d13d9958c7b5b4d">apop_vector_normalize</a>((gsl_vector *)in, .normalization_type=<span class="charliteral">&#39;s&#39;</span>);</div>
<div class="line">    gsl_vector_mul(normed, normed);</div>
<div class="line">    <span class="keywordflow">for</span>(<span class="keywordtype">size_t</span> i=0;i&lt; normed-&gt;size; </div>
<div class="line">            sum +=gsl_vector_get(normed,i++));</div>
<div class="line">    gsl_vector_free(normed);</div>
<div class="line">    <span class="keywordflow">return</span> gsl_cdf_chisq_P(sum, in-&gt;size); </div>
<div class="line">}</div>
</div><!-- fragment --> <pre class="fragment">Or, consider the Rao statistic, 
</pre><p> <img class="formulaInl" alt="${\partial\over \partial\beta}\log L(\beta)'I^{-1}(\beta){\partial\over \partial\beta}\log L(\beta)$" src="form_107.png"/> where <img class="formulaInl" alt="$L$" src="form_108.png"/> is your model's likelihood function and <img class="formulaInl" alt="$I$" src="form_109.png"/> its information matrix. In code:</p>
<div class="fragment"><div class="line"><a class="code" href="structapop__data.html">apop_data</a> * infoinv = <a class="code" href="apop_8h.html#a0f2784b69175d1ecd3833d228d2294e7">apop_model_numerical_covariance</a>(data, your_model);</div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> * score;</div>
<div class="line"><a class="code" href="group__models.html#ga228b013bc80f308900882312cab93eb3">apop_score</a>(data, score-&gt;vector, your_model);</div>
<div class="line"><a class="code" href="structapop__data.html">apop_data</a> * stat = <a class="code" href="group__linear__algebra.html#gaf2d13e1b1f20f97e0772a0be41dc430e">apop_dot</a>(<a class="code" href="group__linear__algebra.html#gaf2d13e1b1f20f97e0772a0be41dc430e">apop_dot</a>(score, infoinv), score);</div>
</div><!-- fragment --><p>Given the correct assumptions, this is <img class="formulaInl" alt="$\sim \chi^2_m$" src="form_110.png"/>, where <img class="formulaInl" alt="$m$" src="form_111.png"/> is the dimension of <img class="formulaInl" alt="$\beta$" src="form_5.png"/>, so the odds of a Type I error given the model is:</p>
<div class="fragment"><div class="line"><span class="keywordtype">double</span> p_value = <a class="code" href="apop_8h.html#a650d1ecf6f78c3bdb85c8fce71dfae05">apop_test</a>(stat, <span class="stringliteral">&quot;chi squared&quot;</span>, beta-&gt;size);</div>
</div><!-- fragment --><p></div>
<p><h2><a class="anchor" name="Histosec"><div class="trigger" onClick="showBranch('Histosecd');swapFolder('Histosecf')"><img src="right.png" border="0" id="Histosecf" alt="pip"> Empirical distributions and PMFs (probability mass functions)</div></a></h2><div class="branch" id="Histosecd">
<p>The <a class="el" href="group__models.html#gaca8f3323c57e0223a9f3c0f991c9760e">apop_pmf</a> model wraps a <a class="el" href="gentle.html#apop_data">apop_data</a> set so it can be read as an empirical model, with a likelihoood function (equal to the associated weight for observed values and zero for unobserved values), a random number generator (which simply makes weighted random draws from the data), and so on. Setting it up is a model estimation from data like any other, done via <a class="el" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(<code>your_data</code>, <a class="el" href="group__models.html#gaca8f3323c57e0223a9f3c0f991c9760e">apop_pmf</a>).</p>
<p>You have the option of cleaning up the data before turning it into a PMF. For example...</p>
<div class="fragment"><div class="line"><a class="code" href="apop_8h.html#acc90578a8086a2e6f7b1575300260f08">apop_data_pmf_compress</a>(your_data);          <span class="comment">//remove duplicates</span></div>
<div class="line"><a class="code" href="apop_8h.html#acefe4dee819c090e57ac69d4d8bdcfd0">apop_data_sort</a>(your_data);</div>
<div class="line"><a class="code" href="apop_8h.html#a194ad92b1bc632ba1d13d9958c7b5b4d">apop_vector_normalize</a>(your_data-&gt;weights);  <span class="comment">//weights sum to one</span></div>
<div class="line"><a class="code" href="structapop__model.html">apop_model</a> *a_pmf = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(your_data, <a class="code" href="group__models.html#gaca8f3323c57e0223a9f3c0f991c9760e">apop_pmf</a>);</div>
</div><!-- fragment --><p>These are largely optional.</p>
<ul>
<li>The CDF is calculated based on the percent of the weights between the zeroth row of the PMF and the row specified. This generally makes more sense after <a class="el" href="apop__sort_8c.html#acefe4dee819c090e57ac69d4d8bdcfd0">apop_data_sort</a>. </li>
<li>Compression produces a corresponding improvement in efficiency when calculating CDFs, but is otherwise not necessary. </li>
<li>Sorting or normalizing is not necessary for making draws or getting a likelihood or log likelihood.</li>
</ul>
<p>It is the weights vector that holds the density represented by each row; the rest of the row represents the coordinates of that density. If the input data set has no <code>weights</code> segment, then I assume that all rows have equal weight.</p>
<p>Most models have a <code>parameters</code> <a class="el" href="gentle.html#apop_data">apop_data</a> set that is filled when you call <a class="el" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>. For a PMF model, the <code>parameters</code> are <code>NULL</code>, and the <code>data</code> itself is used for calculation. Therefore, modifying the data post-estimation can break some internal settings set during estimation. If you modify the data, throw away any existing PMFs (<code>apop_model_free</code>) and re-estimate a new one.</p>
<p>Using <a class="el" href="apop_8h.html#acc90578a8086a2e6f7b1575300260f08">apop_data_pmf_compress</a> puts the data into one bin for each unique value in the data set. You may instead want bins of fixed with, in the style of a histogram, which you can get via <a class="el" href="apop__hist_8c.html#a02cef4243593cf905fa54cb48d8b04c2">apop_data_to_bins</a>. It requires a bin specification. If you send a <code>NULL</code> binspec, then the offset is zero and the bin size is big enough to ensure that there are <img class="formulaInl" alt="$\sqrt(N)$" src="form_17.png"/> bins from minimum to maximum. There are other preferred formulæ for bin widths that minimize MSE, which might be added as a future extension. The binspec will be added as a page to the data set, named <code>"&lt;binspec&gt;"</code>. See the <a class="el" href="apop__hist_8c.html#a02cef4243593cf905fa54cb48d8b04c2">apop_data_to_bins</a> documentation on how to write a custom bin spec.</p>
<p>There are a few ways of testing the claim that one distribution equals another, typically an empirical PMF versus a smooth theoretical distribution. In both cases, you will need two distributions based on the same binspec.</p>
<p>For example, if you do not have a prior binspec in mind, then you can use the one generated by the first call to the histogram binning function to make sure that the second data set is in sync:</p>
<div class="fragment"><div class="line"><a class="code" href="apop_8h.html#a02cef4243593cf905fa54cb48d8b04c2">apop_data_to_bins</a>(first_set, NULL);</div>
<div class="line"><a class="code" href="apop_8h.html#a02cef4243593cf905fa54cb48d8b04c2">apop_data_to_bins</a>(second_set, <a class="code" href="apop_8h.html#a044771db76f0afb410a8d4f05ac51ca3">apop_data_get_page</a>(first_set, <span class="stringliteral">&quot;&lt;binspec&gt;&quot;</span>));</div>
</div><!-- fragment --><p>You can use <a class="el" href="group__histograms.html#ga98ccaa9de7cb16e60a435c56eaf3400d">apop_test_kolmogorov</a> or <a class="el" href="group__histograms.html#gad999220164049ce7b22d2cd11bf74c26">apop_histograms_test_goodness_of_fit</a> to generate the appropriate statistics from the pairs of bins.</p>
<p>Kernel density estimation will produce a smoothed PDF. See <a class="el" href="group__models.html#gaded18b6bf8bd63a53f7b72ed5f27fdb5">apop_kernel_density</a> for details. Or, use <a class="el" href="apop__smoothing_8c.html#a9f236ac6d9bff6835c185b9f098d46d3">apop_vector_moving_average</a> for a simpler smoothing method.</p>
<ul>
<li><a class="el" href="apop_8h.html#acc90578a8086a2e6f7b1575300260f08">apop_data_pmf_compress()</a> : merge together redundant rows in a data set before calling <a class="el" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(<code>your_data</code>, <a class="el" href="group__models.html#gaca8f3323c57e0223a9f3c0f991c9760e">apop_pmf</a>); optional. </li>
<li><a class="el" href="apop_8h.html#acfa62ab9157d4d92338a29093c7d1b30">apop_vector_moving_average()</a> : smooth a vector (e.g., your_pmf-&gt;data-&gt;weights) via moving average. </li>
<li><a class="el" href="group__histograms.html#gad999220164049ce7b22d2cd11bf74c26">apop_histograms_test_goodness_of_fit()</a> : goodness-of-fit via <img class="formulaInl" alt="$\chi^2$" src="form_18.png"/> statistic </li>
<li><a class="el" href="group__histograms.html#ga98ccaa9de7cb16e60a435c56eaf3400d">apop_test_kolmogorov()</a> : goodness-of-fit via Kolmogorov-Smirnov statistic</li>
</ul>
<p></div>
<p><h2><a class="anchor" name="Maxi"><div class="trigger" onClick="showBranch('Maxid');swapFolder('Maxif')"><img src="right.png" border="0" id="Maxif" alt="pip"> Maximum likelihood methods</div></a></h2><div class="branch" id="Maxid">
<p>This section includes some notes on the maximum likelihood routine. As in the section on writing models above, if a model has a <code>p</code> or <code>log_likelihood</code> method but no <code>estimate</code> method, then calling <code>apop_estimate(your_data, your_model)</code> executes the default estimation routine of maximum likelihood.</p>
<p>If you are a not a statistician, then there are a few things you will need to keep in mind:</p>
<ul>
<li>Physicists, pure mathematicians, and the GSL minimize; economists, statisticians, and Apophenia maximize. If you are doing a minimization, be sure that your function returns minus the objective function's value.</li>
</ul>
<ul>
<li>The overall setup is about estimating the parameters of a model with data. The user provides a data set and an unparameterized model, and the system tries parameterized models until one of them is found to be optimal. The data is fixed. The parameters make sense only in the context of a model, so the optimization tries a series of parameterized models, searching for the one that is most likely. In a non-stats setting, you will probably have <code>NULL</code> data, and will be optimizing the parameters internal to the model.</li>
</ul>
<ul>
<li>Because the unit of analysis is a parameterized model, not just parameters, you need to have an <a class="el" href="structapop__model.html">apop_model</a> wrapping your objective function. Here is a typical sort of function that one would maximize; it is Rosenbrock's banana function, <img class="formulaInl" alt="$(1-x)^2+ s(y - x^2)^2$" src="form_112.png"/>, where the scaling factor <img class="formulaInl" alt="$s$" src="form_113.png"/> is fixed ahead of time, say at 100.</li>
</ul>
<div class="fragment"><div class="line"><span class="keyword">typedef</span> <span class="keyword">struct </span>{</div>
<div class="line">        <span class="keywordtype">double</span> scaling;</div>
<div class="line">} <a class="code" href="structcoeff__struct.html">coeff_struct</a>;</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">double</span> banana (<span class="keywordtype">double</span> *params, <a class="code" href="structcoeff__struct.html">coeff_struct</a> *in){</div>
<div class="line">        <span class="keywordflow">return</span> (gsl_pow_2(1-params[0]) +</div>
<div class="line">                in-&gt;scaling*gsl_pow_2(params[1]-gsl_pow_2(params[0])));</div>
<div class="line">}</div>
</div><!-- fragment --><p>The function returns a single number to be minimized. You will need to write an <a class="el" href="structapop__model.html">apop_model</a> to send to the optimizer, which is a two step process: write a log likelihood function wrapping the real objective function, and a model that uses that log likelihood. Here they are:</p>
<div class="fragment"><div class="line"><span class="keywordtype">double</span> ll (<a class="code" href="structapop__data.html">apop_data</a> *d, <a class="code" href="structapop__model.html">apop_model</a> *in){</div>
<div class="line">    <span class="keywordflow">return</span> - banana(in-&gt;<a class="code" href="structapop__model.html#ae016a9e13725e84a1188bc81c8f09e45">parameters</a>-&gt;vector-&gt;data, in-&gt;<a class="code" href="structapop__model.html#ac6ae82f82cf5dbc3ae21919f9978cb7b">more</a>);</div>
<div class="line">}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">int</span> main(){</div>
<div class="line">    <a class="code" href="structcoeff__struct.html">coeff_struct</a> co = {.scaling=100};</div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> b = {<span class="stringliteral">&quot;Bananas!&quot;</span>, .<a class="code" href="structapop__model.html#a25bf144576642b48097e764318a6ce28">log_likelihood</a>= ll, .vsize=2,</div>
<div class="line">                                .more = &amp;co, .more_size=<span class="keyword">sizeof</span>(<a class="code" href="structcoeff__struct.html">coeff_struct</a>)};</div>
</div><!-- fragment --><p>The <code>vsize=2</code> specified that your parameters are a vector of size two, which means that <code>in-&gt;parameters-&gt;vector-&gt;data</code> is the list of <code>double</code>s that you should send to <code>banana</code>. The <code>more</code> element of the structure is designed to hold any arbitrary structure; if you use it, you will also need to use the <code>more_size</code> element, as above.</p>
<ul>
<li>Statisticians want the covariance and basic tests about the parameters. If you just want the optimal value, then adding this line will shut off all auxiliary calculations: <div class="fragment"><div class="line"><a class="code" href="group__settings.html#ga8fb1877a3cc29edd685a68dd6b4a35dc">Apop_settings_add_group</a>(your_model, apop_parts_wanted);</div>
</div><!-- fragment --> See the documentation for <a class="el" href="structapop__parts__wanted__settings.html">apop_parts_wanted_settings</a> for details about how this works. It can also offer quite the speedup: especially for high-dimensional problems, finding the covariance matrix without any information can take dozens of evaluations of the objective funtion for each evaluation that is part of the search itself.</li>
</ul>
<ul>
<li>MLEs have an especially large number of parameter tweaks that could be made; see the section on MLE settings above or the <a class="el" href="structapop__mle__settings.html">apop_mle_settings</a> page.</li>
</ul>
<ul>
<li>Putting it all together, here is a full program to minimize Rosenbrock's banana function. There are some extras: it uses two methods (notice how easy it is to re-run an estimation with an alternate method, but the syntax for modifying a setting differs from the initialization syntax) and checks that the results are accurate.</li>
</ul>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="apop_8h.html">apop.h</a>&gt;</span></div>
<div class="line"></div>
<div class="line"><span class="keyword">typedef</span> <span class="keyword">struct </span>{</div>
<div class="line">    <span class="keywordtype">double</span> scaling;</div>
<div class="line">} <a class="code" href="structcoeff__struct.html">coeff_struct</a>;</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">long</span> <span class="keywordtype">double</span> banana (<span class="keywordtype">double</span> *params, <a class="code" href="structcoeff__struct.html">coeff_struct</a> *in){</div>
<div class="line">    <span class="keywordflow">return</span> (gsl_pow_2(1-params[0]) </div>
<div class="line">               + in-&gt;scaling*gsl_pow_2(params[1]-gsl_pow_2(params[0])));</div>
<div class="line">}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">long</span> <span class="keywordtype">double</span> ll (<a class="code" href="structapop__data.html">apop_data</a> *d, <a class="code" href="structapop__model.html">apop_model</a> *in){</div>
<div class="line">    <span class="keywordflow">return</span> - banana(in-&gt;<a class="code" href="structapop__model.html#ae016a9e13725e84a1188bc81c8f09e45">parameters</a>-&gt;vector-&gt;data, in-&gt;<a class="code" href="structapop__model.html#ac6ae82f82cf5dbc3ae21919f9978cb7b">more</a>);</div>
<div class="line">}</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">int</span> main(){</div>
<div class="line">    <a class="code" href="structcoeff__struct.html">coeff_struct</a> co = {.scaling=100};</div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *b = &amp;(<a class="code" href="apop_8h.html#afe86ae10fc82d219906211e4f88e4cf9">apop_model</a>) {<span class="stringliteral">&quot;¡Bananas!&quot;</span>, .<a class="code" href="structapop__model.html#a25bf144576642b48097e764318a6ce28">log_likelihood</a>= ll,</div>
<div class="line">                     .vsize=2, .more = &amp;co, .more_size=<span class="keyword">sizeof</span>(<a class="code" href="structcoeff__struct.html">coeff_struct</a>)};</div>
<div class="line">    Apop_model_add_group(b, apop_mle, .verbose=<span class="charliteral">&#39;y&#39;</span>, .method=<span class="stringliteral">&quot;NM simplex&quot;</span>);</div>
<div class="line">    Apop_model_add_group(b, apop_parts_wanted);</div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *e1 = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(NULL, b);</div>
<div class="line">    <a class="code" href="group__output.html#gab7bf9adff9b97c60c5839ae72ba4fb7f">apop_model_print</a>(e1, NULL);</div>
<div class="line"></div>
<div class="line">    <a class="code" href="group__settings.html#ga8659282b6003f7c2adac6f4baad5fae4">Apop_settings_set</a>(b, apop_mle, method, <span class="stringliteral">&quot;BFGS cg&quot;</span>);</div>
<div class="line">    <a class="code" href="structapop__model.html">apop_model</a> *e2 = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(NULL, b);</div>
<div class="line">    <a class="code" href="group__output.html#gab7bf9adff9b97c60c5839ae72ba4fb7f">apop_model_print</a>(e2, NULL);</div>
<div class="line"></div>
<div class="line">    gsl_vector *one = <a class="code" href="apop_8h.html#af9b051e2fbe0e473dddeabfe4932ce0c">apop_vector_fill</a>(gsl_vector_alloc(2), 1, 1);</div>
<div class="line">    assert(<a class="code" href="group__convenience__fns.html#ga44a77f5e6867c627fd313f3f6e252ff8">apop_vector_distance</a>(e1-&gt;<a class="code" href="structapop__model.html#ae016a9e13725e84a1188bc81c8f09e45">parameters</a>-&gt;vector, one) &lt; 1e-2);</div>
<div class="line">    assert(<a class="code" href="group__convenience__fns.html#ga44a77f5e6867c627fd313f3f6e252ff8">apop_vector_distance</a>(e2-&gt;<a class="code" href="structapop__model.html#ae016a9e13725e84a1188bc81c8f09e45">parameters</a>-&gt;vector, one) &lt; 1e-2);</div>
<div class="line">}</div>
</div><!-- fragment --><ul>
<li>If you would like to see what the optimizer did, add <code>.want_path='y'</code> to the settings group, then get the <code>path</code> element from the settings group:</li>
</ul>
<div class="fragment"><div class="line"><a class="code" href="group__settings.html#ga8fb1877a3cc29edd685a68dd6b4a35dc">Apop_settings_add_group</a>(your_model, apop_mle, .want_path=<span class="charliteral">&#39;y&#39;</span>);</div>
<div class="line"><a class="code" href="structapop__model.html">apop_model</a> *out = <a class="code" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a>(your_data, your_model);</div>
<div class="line"><a class="code" href="group__output.html#ga9a30c437ec11a11e37e0db7f82386275">apop_data_show</a>(<a class="code" href="group__settings.html#ga4ba53402807c9e6b82254fb552029501">Apop_settings_get</a>(out, apop_mle, path));</div>
</div><!-- fragment --><p><h2><a class="anchor" name="constr"><div class="trigger" onClick="showBranch('constrd');swapFolder('constrf')"><img src="right.png" border="0" id="constrf" alt="pip"> Setting Constraints</div></a></h2><div class="branch" id="constrd">
<p>The problem is that the parameters of a function must not take on certain values, either because the function is undefined for those values or because parameters with certain values would not fit the real-world problem.</p>
<p>The solution is to rewrite the function being maximized such that the function is continuous at the constraint boundary but takes a steep downward slope. The unconstrained maximization routines will be able to search a continuous function but will never return a solution that falls beyond the parameter limits.</p>
<p>If you give it a likelihood function with no regard to constraints plus an array of constraints, <a class="el" href="group__mle.html#gae9f092d1786034dd2ab22c63de5c955c">apop_maximum_likelihood</a> will combine them to a function that fits the above description and search accordingly.</p>
<p>A constraint function must do three things: </p><ul>
<li>It must check the constraint, and if the constraint does not bind (i.e. the parameter values are OK), then it must return zero. </li>
<li>If the constraint does bind, it must return a penalty, that indicates how far off the parameter is from meeting the constraint. </li>
<li>if the constraint does bind, it must set a return vector that the likelihood function can take as a valid input. The penalty at this returned value must be zero.</li>
</ul>
<p>The idea is that if the constraint returns zero, the log likelihood function will return the log likelihood as usual, and if not, it will return the log likelihood at the constraint's return vector minus the penalty. To give a concrete example, here is a constraint function that will ensure that both parameters of a two-dimensional input are both greater than zero. As with the constraints for many of the models that ship with Apophenia, it is a wrapper for <a class="el" href="apop__linear__constraint_8c.html#ad490e5b31d1bbc99223c2cbfac7c1782">apop_linear_constraint</a>.</p>
<div class="fragment"><div class="line"><span class="keyword">static</span> <span class="keywordtype">long</span> <span class="keywordtype">double</span> beta_zero_greater_than_x_constraint(<a class="code" href="structapop__data.html">apop_data</a> *data, <a class="code" href="structapop__model.html">apop_model</a> *v){</div>
<div class="line">    <span class="comment">//constraint is 0 &lt; beta_2</span></div>
<div class="line">    <span class="keyword">static</span> <a class="code" href="structapop__data.html">apop_data</a> *constraint = NULL;</div>
<div class="line">    <span class="keywordflow">if</span> (!constraint) constraint= <a class="code" href="apop_8h.html#acb6b2c53478f9db902ba0340ba499819">apop_data_falloc</a>((1,1,1), 0, 1);</div>
<div class="line">    <span class="keywordflow">return</span> <a class="code" href="apop_8h.html#ad490e5b31d1bbc99223c2cbfac7c1782">apop_linear_constraint</a>(v-&gt;<a class="code" href="structapop__model.html#ae016a9e13725e84a1188bc81c8f09e45">parameters</a>-&gt;vector, constraint, 1e-3);</div>
<div class="line">}</div>
</div><!-- fragment --><ul>
<li><a class="el" href="apop_8h.html#ad490e5b31d1bbc99223c2cbfac7c1782">apop_linear_constraint()</a></li>
</ul>
<p></div>
<ul>
<li><a class="el" href="group__mle.html#ga6d5e39c00012de5b51cd49ce1ffd31fe">apop_estimate_restart()</a>: Restarting an MLE with different settings can improve results. </li>
<li><a class="el" href="group__mle.html#gae9f092d1786034dd2ab22c63de5c955c">apop_maximum_likelihood()</a>: Rarely used. If a model has no <code>estimate</code> element, just call <a class="el" href="group__models.html#ga2c0598e5aca68949939a189bd4ce24ca">apop_estimate</a> to run an MLE. </li>
<li><a class="el" href="apop_8h.html#a0f2784b69175d1ecd3833d228d2294e7">apop_model_numerical_covariance()</a> </li>
<li><a class="el" href="group__linear__algebra.html#ga1b6ecb8702153658c29369dc2e7cfba9">apop_numerical_gradient()</a></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="Miss"><div class="trigger" onClick="showBranch('Missd');swapFolder('Missf')"><img src="right.png" border="0" id="Missf" alt="pip"> Missing data</div></a></h2><div class="branch" id="Missd">
<ul>
<li><a class="el" href="apop_8h.html#a0316d98bbeb6ae13532a8c071019dbe6">apop_data_listwise_delete()</a> </li>
<li><a class="el" href="apop_8h.html#a2a184259c139f1fdd970ef6b43f6274e">apop_ml_impute()</a></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="Legi"><div class="trigger" onClick="showBranch('Legid');swapFolder('Legif')"><img src="right.png" border="0" id="Legif" alt="pip"> Legible output</div></a></h2><div class="branch" id="Legid">
<p>The output routines handle four sinks for your output. There is a global variable that you can use for small projects where all data will go to the same place.</p>
<div class="fragment"><div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_type = <span class="charliteral">&#39;s&#39;</span>; <span class="comment">//Stdout</span></div>
<div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_type = <span class="charliteral">&#39;f&#39;</span>; <span class="comment">//named file</span></div>
<div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_type = <span class="charliteral">&#39;p&#39;</span>; <span class="comment">//a pipe or already-opened file</span></div>
<div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_type = <span class="charliteral">&#39;d&#39;</span>; <span class="comment">//the database</span></div>
</div><!-- fragment --><p>You can also set the output type, the name of the output file or table, and other options via arguments to individual calls to output functions. See <a class="el" href="apop__output_8c.html#a4748bf59f5441cd27e5b93056c7ec1c7">apop_prep_output</a> for the list of options.</p>
<p>C makes minimal distinction between pipes and files, so you can set a pipe or file as output and send all output there until further notice:</p>
<div class="fragment"><div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_type = <span class="charliteral">&#39;p&#39;</span>;</div>
<div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_pipe = popen(<span class="stringliteral">&quot;gnuplot&quot;</span>, <span class="stringliteral">&quot;w&quot;</span>);</div>
<div class="line">apop_plot_lattice(...); <span class="comment">//see https://github.com/b-k/Apophenia/wiki/gnuplot_snippets</span></div>
<div class="line">fclose(<a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_pipe);</div>
<div class="line"><a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_pipe = fopen(<span class="stringliteral">&quot;newfile&quot;</span>, <span class="stringliteral">&quot;w&quot;</span>);</div>
<div class="line"><a class="code" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print</a>(set1);</div>
<div class="line">fprintf(<a class="code" href="apop_8h.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a>.output_pipe, <span class="stringliteral">&quot;\nNow set 2:\n&quot;</span>);</div>
<div class="line"><a class="code" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print</a>(set2);</div>
</div><!-- fragment --><p>Continuing the example, you can always override the global data with a specific request: </p><div class="fragment"><div class="line"><a class="code" href="group__apop__print.html#ga6e68deb42339cf6068ab6e741237918d">apop_vector_print</a>(v, <span class="stringliteral">&quot;vectorfile&quot;</span>); <span class="comment">//put vectors in a separate file</span></div>
<div class="line"><a class="code" href="group__apop__print.html#ga50915e9992a7cc880807cc266e9f5166">apop_matrix_print</a>(m, <span class="stringliteral">&quot;matrix_table&quot;</span>, .output_type = <span class="charliteral">&#39;d&#39;</span>); <span class="comment">//write to the db</span></div>
<div class="line"><a class="code" href="group__apop__print.html#ga50915e9992a7cc880807cc266e9f5166">apop_matrix_print</a>(m, .output_pipe = stdout);  <span class="comment">//now show the same matrix on screen</span></div>
</div><!-- fragment --><p>I will first look to the input file name, then the input pipe, then the global <code>output_pipe</code>, in that order, to determine to where I should write. Some combinations (like output type = <code>'d'</code> and only a pipe) don't make sense, and I'll try to warn you about those.</p>
<p>What if you have too much output and would like to use a pager, like <code>less</code> or <code>more</code>? In C and POSIX terminology, you're asking to pipe your output to a paging program. Here is the form: </p><div class="fragment"><div class="line">FILE *lesspipe = popen(<span class="stringliteral">&quot;less&quot;</span>, <span class="stringliteral">&quot;w&quot;</span>);</div>
<div class="line">assert(lesspipe);</div>
<div class="line"><a class="code" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print</a>(your_data_set, .output_pipe=lesspipe);</div>
<div class="line">pclose(lesspipe);</div>
</div><!-- fragment --><p> <code>popen</code> will search your usual program path for <code>less</code>, so you don't have to give a full path.</p>
<ul>
<li><a class="el" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print()</a> </li>
<li><a class="el" href="group__apop__print.html#ga50915e9992a7cc880807cc266e9f5166">apop_matrix_print()</a> </li>
<li><a class="el" href="group__apop__print.html#ga6e68deb42339cf6068ab6e741237918d">apop_vector_print()</a> </li>
<li><a class="el" href="group__output.html#ga9a30c437ec11a11e37e0db7f82386275">apop_data_show()</a> : alias for <a class="el" href="group__apop__print.html#gab45a848e5dbcaf82b5b95cd03f7b396a">apop_data_print</a> limited to <code>stdout</code>.</li>
</ul>
<p></div>
<p><h2><a class="anchor" name="moreasst"><div class="trigger" onClick="showBranch('moreasstd');swapFolder('moreasstf')"><img src="right.png" border="0" id="moreasstf" alt="pip"> Assorted</div></a></h2><div class="branch" id="moreasstd">
<p>A few more descriptive methods:</p>
<ul>
<li><a class="el" href="group__linear__algebra.html#gaa1715c468336800a21ca86971323ae3d">apop_matrix_pca</a> : Principal component analysis </li>
<li><a class="el" href="apop__tests_8c.html#a1309242dbb7f148916c71a9a70c6b64c">apop_anova</a> : One-way or two-way ANOVA tables </li>
<li><a class="el" href="apop_8h.html#ac1bae5dffbbd4bc82c7369211c36f934">apop_rake</a> : Iterative proportional fitting on large, sparse tables</li>
</ul>
<p>General utilities:</p>
<ul>
<li><a class="el" href="apop_8h.html#abd17601e642688d033f62d346f3dace6">Apop_stopif</a> : Apophenia's error-handling and warning-printing macro. </li>
<li><a class="el" href="apop__db_8c.html#a7975faa07196cf463ec261ff0ddc3ccc">apop_opts</a> : the global options </li>
<li><a class="el" href="apop_8h.html#aa9ec5e7ed223c09bec044ced54f24fad">apop_regex()</a> : friendlier front-end for POSIX-standard regular expression searching and pulling matches into a <a class="el" href="gentle.html#apop_data">apop_data</a> set. </li>
<li><a class="el" href="apop_8h.html#a6270ad55c566140e0d7962cb4d5daacc">apop_text_paste()</a>: produce a single string from a grid of text </li>
<li><a class="el" href="apop__asst_8c.html#a153362f8a4c7f64e0c8dad893edadea4">apop_system()</a></li>
</ul>
<p>Math utilities:</p>
<ul>
<li><a class="el" href="apop_8h.html#afc7707d44884aa63c298510a0662956a">apop_matrix_is_positive_semidefinite()</a> </li>
<li><a class="el" href="apop_8h.html#af1fc93322ffcb4e20cfc0b9af9e2aea0">apop_matrix_to_positive_semidefinite()</a> </li>
<li><a class="el" href="apop_8h.html#a757dd4278aac9967f9e62c3ef3c8fc69">apop_generalized_harmonic()</a> </li>
<li><a class="el" href="apop_8h.html#a2c76ad8792959b6605402a1d71a1b6c9">apop_multivariate_gamma()</a> </li>
<li><a class="el" href="apop_8h.html#a180b231acc3c64d07664ed2b1a8ad0e1">apop_multivariate_lngamma()</a> </li>
<li><a class="el" href="group__convenience__fns.html#gada044cf02135422a9f9292b6054c86dd">apop_rng_alloc()</a></li>
</ul>
<p></div>
<p><h2><a class="anchor" name="links"><div class="trigger" onClick="showBranch('linksd');swapFolder('linksf')"><img src="right.png" border="0" id="linksf" alt="pip"> Further references</div></a></h2><div class="branch" id="linksd">
<p>For your convenience, here are links to some other libraries you are probably using.</p>
<ul>
<li><a href="http://www.gnu.org/software/libc/manual/html_node/index.html">The standard C library</a> </li>
<li><a href="http://www.gnu.org/software/gsl/manual/html_node/index.html">The GSL documentation</a>, and <a href="http://www.gnu.org/software/gsl/manual/html_node/Function-Index.html">its index</a> </li>
<li><a href="http://sqlite.org/lang.html">SQL understood by SQLite</a> </li>
</ul>
</div></div><!-- contents -->
<p><p>
<div class="tiny">Autogenerated by doxygen on Mon Apr 20 2015.</div></body></html>
